<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Intro to Linux" />
<meta property="og:type" content="book" />
<meta property="og:image" content="/figures/NEOF.png" />
<meta property="og:description" content="NEOF book for the Introduction to command line bioinformatics workshop" />


<meta name="author" content="Matthew R. Gemmell" />

<meta name="date" content="2024-02-23" />


<meta name="description" content="NEOF book for the Introduction to command line bioinformatics workshop">

<title>Intro to Linux</title>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<!--bookdown:toc:end-->
<!--bookdown:toc:start-->
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="header">
<h1 class="title">Intro to Linux</h1>
<h4 class="author"><em>Matthew R. Gemmell</em></h4>
<h4 class="date"><em>2024-02-23</em></h4>
</div>
<p><img src="figures/NEOF.png" width="30%" style="display: block; margin: auto;" /></p>
<div id="intro" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p><img src="figures/linux_beginner.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Bioinformatics is an increasingly important skill for biological scientists. Many bioinformatic tools can only be run on Linux based operating systems. This course aims to introduce you to Linux and is aimed at beginners and novices to the command line.</p>
<p>Sessions will start with a brief presentation followed by self-paced computer practicals guided by an online interactive book. The book will contain theory, practice code, and exercises. Multiple choice questions will help reinforce what you have learnt throughout the book.</p>
<p>At the end of the course learners will be able to:</p>
<ul>
<li>Explain how the user, shell, and kernel interact with each other in the Linux OS.</li>
<li>Navigate &amp; manipulate directories in the Linux environment.</li>
<li>List and view the contents of directories.</li>
<li>Carry out a variety of commands with files including printing them to terminal.</li>
<li>Utilise the text editor nano to create, edit, and save files.</li>
<li>Understand the structure of fastq files.</li>
</ul>
<p>There are additional materials in chapters 12-14 which include some advanced Linux commands and introductions to other programming languages used by bioinformaticians.</p>
<p>Commands are in the following font, colour, and box.
They should be run in the command line.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;This is a command example&quot;</span> </span></code></pre></div>
<p>In some chapters there are video walk-throughs.
These are optional and give you a guided visual walk-through of running the practice code in the chapter.
These are found in an expandable box at the start of the chapters 4,5,6, and 8.</p>
<p>Additionally, please use the <a href="#cheatsheet">cheatsheet</a> as a reminder of all the commands you will be learning.</p>
<div id="table-of-contents" class="section level2 unnumbered">
<h2>Table of contents</h2>
<table>
<colgroup>
<col width="51%" />
<col width="48%" />
</colgroup>
<tbody>
<tr class="odd">
<td align="center"><p><a href="#linuxintro"><strong>What is Linux?</strong></a></p>
<p><a href="#linuxintro"><img src="figures/linux_beginner.png" width="142" /></a></p></td>
<td align="center"><p><a href="#cluster"><strong>Logging in to our teaching VNC</strong></a></p>
<p><a href="#cluster"><img src="figures/start.png" width="189" /></a></p></td>
</tr>
<tr class="even">
<td align="center"><p><a href="#dirsandfiles"><strong>Directories and files</strong></a></p>
<p><a href="#dirsandfiles"><img src="figures/directory.png" width="125" /></a></p></td>
<td align="center"><p><a href="#tipsandtricks"><strong>Tips and tricks</strong></a></p>
<p><a href="#tipsandtricks"><img src="figures/skateboard_trick.png" width="302" /></a></p></td>
</tr>
<tr class="odd">
<td align="center"><p><a href="#manipulatingdirectories"><strong>Manipulating Directories</strong></a></p>
<p><a href="#manipulatingdirectories"><img src="figures/transform.png" width="150" /></a></p></td>
<td align="center"><p><a href="#filereadingandprocessing"><strong>File reading and processing</strong></a></p>
<p><a href="#filereadingandprocessing"><img src="figures/files.png" width="205" /></a></p></td>
</tr>
<tr class="even">
<td align="center"><p><a href="#advancedlinux"><strong>Advanced Linux</strong></a></p>
<p><a href="#advancedlinux"><img src="figures/linux_intermdiary.png" width="142" /></a></p></td>
<td align="center"><p><a href="#bfxlanguages"><strong>Other Bioinformatics programming languages</strong></a></p>
<p><a href="#bfxlanguages"><img src="figures/languages.png" width="159" /></a></p></td>
</tr>
<tr class="odd">
<td align="center"><p><a href="#cheatsheet"><strong>Appendix</strong></a></p>
<p><a href="#cheatsheet"><img src="figures/cheatsheet.png" width="150" /></a></p></td>
<td align="center"></td>
</tr>
</tbody>
</table>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" alt="Creative Commons Licence" style="border-width:0"/></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.</p>

</div>
</div>



<div id="linuxintro" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Linux</h1>
<p><img src="figures/linux_beginner.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Linux is a multitasking, multiuser Unix-like computer operating system (OS). Linux can run many different applications (multitasking) and it can be used by many different people (multiuser) on the same computer at the same time.</p>
<p>It is utilised by many programmers, including Bioinformaticians. It is a relatively easy OS to run commands and develop software for. The vast majority of programs and tools for computational analysis of biological data will work in Linux.</p>
<p>There are three parts of the Linux OS:</p>
<ul>
<li><strong>The kernel</strong>: This is the hub of the operating systems. This is the “behind the scenes” part of the OS which allocates time and memory to programs. This controls the hardware.</li>
<li><strong>The shell</strong>: This acts as the interface between the user and the kernel. When a user runs commands, the shell will interpret these commands for the kernel. The shell can be any program that constitutes the user interface e.g. command line, internet browser, start menu etc.</li>
<li><strong>Programs</strong>: Programs allow the OS to perform specific tasks. Examples of programs include Internet browsers, genome assembly tools, text editors etc.</li>
</ul>
<p><img src="figures/linux_user_to_kernel.png" width="80%" style="display: block; margin: auto;" /></p>
<p>It is important to learn the Linux language so you can run commands on the command line. This is because:</p>
<ul>
<li>Many bioinformatic tools do not have a graphical user interface (gui) and so must be run on the command line.</li>
<li>There are many powerful commands that can be run on the Linux command line.</li>
<li>It is quicker and more reproducible to run commands through a shell than through a gui.</li>
</ul>

</div>
<div id="starting" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Starting</h1>
<p><img src="figures/start.png" width="20%" style="display: block; margin: auto;" /></p>
<div id="cluster" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Logon instructions</h2>
<p>For this workshop we will be using Virtual Network Computing (VNC). Connect to the VNC with a browser by using the webVNC link you were sent.</p>
<p>You will now be in a logged-in Linux VNC desktop with two terminals. You will see something as below (there may be only one terminal which is fine). If you do not see something similar please ask for assistance.</p>
<p><img src="figures/logon_pic.png" width="80%" style="display: block; margin: auto;" /></p>
<p>If the VNC is taking up too much/little space of your browser you can use the zoom of your browser to adjust the size. Ensure you can see one whole terminal.</p>
<p>These instructions will not work outside of this workshop. If you would like to install your own Linux OS on your desktop or laptop we would recommend Mint Linux</p>
<p>The following link is a guide to install Mint Linux:<br />
<a href="https://linuxmint-installation-guide.readthedocs.io/en/latest/" class="uri">https://linuxmint-installation-guide.readthedocs.io/en/latest/</a></p>
</div>
<div id="the-terminal-window" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> The Terminal Window</h2>
<p>In our case the terminal window looks like the picture below. We are using the terminal window as our shell to interpret our commands to the kernel. Depending on your system and preferences it may look different.</p>
<p><img src="figures/terminal_window.png" width="80%" style="display: block; margin: auto;" /></p>
<p>Already there is useful information for us on the terminal window.</p>
<ul>
<li><strong>nsc006</strong>: This is the login name, also known as the username. In this case nsc006 is a demonstrator’s account. Your screen should show a different account name which will be your username for the Linux machine/cluster you are logged into.</li>
<li><strong>gauss03</strong>: This is the machine name the user is logged into.</li>
<li><strong>~</strong>: This represents the current directory of the user, or the directory a command was run in. In the Linux OS and others <strong>‘~’</strong> is a shortcut to the user’s home directory.</li>
<li>Everything after the <strong>‘$’</strong> is where commands are typed into the terminal. This is also referred to as the command line.</li>
</ul>
<p>To open a new terminal window, right click on the main screen, choose <code>Applications</code> -&gt; <code>Shell</code> -&gt; <code>bash</code></p>
</div>
<div id="commands" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Commands</h2>
<p>Commands are typed into the terminal and then run by pressing <strong>“enter”</strong>”</p>
<p><img src="figures/enter_symbol.png" width="20%" style="display: block; margin: auto;" /></p>
<p>To run a command select your terminal by clicking on it or using <strong>“alt+shift”</strong>.</p>
<p>Now run the below command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Hello World&quot;</span></span></code></pre></div>
<p>In this case <code>echo</code> is a command that prints out the term “Hello World”.</p>
<p>Now try:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;Bye World&quot;</span></span></code></pre></div>
<p>There are many different Linux commands and we will run through a few. With a large variety it can be hard to remember all the commands and how they work.</p>
<p>Three convenient resources are:</p>
<ol style="list-style-type: decimal">
<li><strong>Search engines</strong> (e.g. Google): There are many forums where people ask for help with command line issues. If you have an issue and are not sure what to do, most likely someone else has had the same issue and asked for help on a forum. The tricky part of this is knowing the specific terminology to use when searching. Forums where people ask bioinformatics questions include SEQanswers, Stack overflow and biostars.</li>
<li><strong>Cheat sheets</strong>: It is never wrong to “cheat” when coding. Cheat sheets with many commands and good descriptions are very useful. Here is a good example of one: <a href="https://files.fosswire.com/2007/08/fwunixref.pdf" class="uri">https://files.fosswire.com/2007/08/fwunixref.pdf</a></li>
<li><strong>Manual pages</strong>: Linux commands have many different parameters and options. If you ever need to figure out what they all are and what they do you can use the <code>man</code> command.</li>
</ol>
<p>E.g. The below command will show the manual page for the echo command:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">man</span> echo</span></code></pre></div>
<p>The below command will show the manual page for the <code>man</code> command:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">man</span> man</span></code></pre></div>
<p><strong>Note</strong>: The <code>man</code> page acts like using the command <strong>less</strong> (we will get into more specifics later). Important notes for now are to use the <strong>arrow keys</strong> to go <strong>up</strong> and <strong>down</strong> the page and press <strong>q</strong> to exit the manual</p>
<p><strong>Note</strong>: There is a cheat sheet at the end of this document with all the commands covered in this practical.</p>

</div>
</div>
<div id="dirsandfiles" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Directories and Files</h1>
<p><img src="figures/directory.png" width="20%" style="display: block; margin: auto;" /></p>
<div class="webex-solution">
<button>
Chapter 4 video walk-through
</button>
<div class="container">
<iframe src="https://www.youtube.com/embed/hJCuJXZQjac" frameborder="0" allowfullscreen class="video">
</iframe>
</div>
</div>
<div id="acquiring-workshop-data" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Acquiring Workshop data</h2>
<p>The first step to carry out is to copy the data for the workshop to your home directory.</p>
<div id="changing-directories" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Changing directories</h3>
<p>Before copying you will change directory to your home directory.</p>
<ul>
<li><code>cd</code> is the command to Change Directory. It is followed by the directory you want to change to.</li>
<li><strong>“~”</strong> represents your home directory.</li>
</ul>
<p>Change directory to your home directory by running the following command in your terminal:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="bu">cd</span> ~ </span></code></pre></div>
<p>In Linux the default of <code>cd</code> is to change directory to your home directory. Therefore the following command will do the same as the above.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span></span></code></pre></div>
<p>To determine your current working directory you can either look at the part of the terminal which displays it or you can use the command <code>pwd</code> (print working directory). Enter the following command:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre></div>
<p>In this case it will not show a <strong>“~”</strong> but the full path of your home directory. E.g. <strong>“/pub14/tea/nsc206/”</strong>.</p>
<ul>
<li>The first <strong>“/”</strong> is the root of the system. Every directory, subdirectory, file and program of the machine is within the root.</li>
<li><strong>“pub14/”</strong>: A directory within the root.</li>
<li><strong>“tea/”</strong>: A subdirectory of <strong>“pub14/”</strong> and a sub-subdirectory of the root (<strong>“/”</strong>).</li>
<li><strong>“nsc206”</strong>: The home directory of user nsc206. It is a subdirectory of <strong>“tea/”</strong> which is a subdirectory of <strong>“pub14/”</strong> which is a subdirectory of the root (<strong>“/”</strong>).</li>
</ul>
</div>
<div id="copying" class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Copying</h3>
<p>Now that we are in our home directory we can copy the data we need to it.</p>
<p>To do this we can use the command <code>cp</code>. This command is followed by the directory/file we want to copy then by the directory we want to copy it to.</p>
<p>To copy a directory we need to add the option <code>-r</code> which means recursively copy this directory and all its contents. Otherwise <code>cp</code> can only be used to copy files.</p>
<p>Use the below command to copy the workshop data to your current directory. The <strong>“.”</strong> refers to your current directory.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> /pub14/tea/nsc206/NEOF/Linux/ .</span></code></pre></div>
<p>Now change directory into your newly copied directory.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="bu">cd</span> Linux</span></code></pre></div>
<p>Print to screen the path of your current working directory.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre></div>
</div>
</div>
<div id="directory-structure" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Directory structure</h2>
<p>You can think of the directory structure in two different ways.</p>
<div id="the-directory-tree" class="section level3" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> The Directory tree</h3>
<p>This starts as the root (<strong>“/”</strong>) which branches out into directories and files. Directories contain files and subdirectories which contain files and subdirectories etc.</p>
<p>Below is an example of visualising the location of the <strong>“Linux”</strong> directory within the user ncs006’s home directory as a tree. This only includes a subset of directories.</p>
<p><img src="figures/linux_tree_structure.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="boxes" class="section level3" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Boxes</h3>
<p>Another analogy to the directory structure is boxes and items. In this case there is one large box that contains all the boxes and items, this is the root (<strong>“/”</strong>). In the root are items and boxes which hold items and boxes etc.</p>
<p>Below is an example of visualising the location of the <strong>“Linux”</strong> directory within the user ncs006’s home directory as boxes. This only includes a subset of directories.</p>
<p><img src="figures/linux_box_structure.png" width="60%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="paths" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Paths</h2>
<p><img src="figures/paths.png" width="20%" style="display: block; margin: auto;" /></p>
<p>On the command line directories and files are referred to by paths e.g. <strong>“/pub14/tea/nsc206/Linux/”</strong> is a path.</p>
<p>Paths are case sensitive. The path <strong>“directory/file.txt”</strong> is different than <strong>“diRectory/File.txt”</strong>.</p>
<p>Spaces should always be avoided in path names. It is highly recommended to <strong>“_”</strong> instead. This is because spaces are used to separate options and parameters in commands.</p>
<p>There are multiple ways to refer to a path. The two main ways are through <strong>absolute paths</strong> and <strong>relative paths</strong>.</p>
<div id="absolute-paths" class="section level3" number="4.3.1">
<h3><span class="header-section-number">4.3.1</span> Absolute paths</h3>
<p>Absolute paths are paths that start from the root e.g.</p>
<ul>
<li><strong>“/pub14/tea/nsc206/Linux/”</strong></li>
<li><strong>“~/Linux/”</strong> (In this case ~ is a shortcut which includes the root)</li>
<li><strong>“/pub14/tea/nsc206/file.txt”</strong></li>
</ul>
</div>
<div id="relative-paths" class="section level3" number="4.3.2">
<h3><span class="header-section-number">4.3.2</span> Relative paths</h3>
<p>Relative paths are paths that are relative to another location besides the root e.g.</p>
<ul>
<li><strong>“.”</strong> (This means the current working directory).</li>
<li><strong>“..”</strong> (This refers to one directory up e.g. if the current directory was <strong>“/pub14/tea/nsc206/”</strong>, the <strong>“..”</strong> directory would be <strong>“/pub14/tea/”</strong>).</li>
<li><strong>“1_directory/”</strong>, would refer to the directory <strong>“1_directory/”</strong> in your current directory</li>
</ul>
</div>
<div id="change-directory-examples" class="section level3" number="4.3.3">
<h3><span class="header-section-number">4.3.3</span> Change directory examples</h3>
<p>Below is a subset of valid methods to change directory into your <strong>“Linux/”</strong> directory</p>
<p><strong>Note</strong>:change <strong>nsc2xx</strong> to your specific user name as shown on the command line prompt.</p>
<p>Method 1</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="bu">cd</span> /pub14/tea/nsc2xx/Linux/</span></code></pre></div>
<p>Method 2</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="bu">cd</span> ~</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="bu">cd</span> Linux/</span></code></pre></div>
<p>Method 3</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="bu">cd</span> /pub14/tea/nsc2xx/Linux/1_directory/</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="bu">cd</span> ..</span></code></pre></div>
<p>Method 4</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">cd</span> /pub14/</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="bu">cd</span> tea/</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="bu">cd</span> nsc2xx/</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="bu">cd</span> Linux/</span></code></pre></div>
</div>
</div>
<div id="listing-directory-content" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> Listing Directory content</h2>
<p><img src="figures/List.png" width="20%" style="display: block; margin: auto;" /></p>
<p>To list the contents (files and directories) within a directory you can use the <code>ls</code> command.</p>
<p>The output of <code>ls</code> will include file names that are coloured black and directory names that are coloured blue in our VNC terminals.</p>
<p>Before carrying out the below commands make sure you are in your <strong>“Linux”</strong> directory with the <code>pwd</code> command.</p>
<p>List the contents in your current directory:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>List the contents in your home directory:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">ls</span> ..</span></code></pre></div>
<p>List the contents in the root directory and put each separate file/directory on a separate line. In the below command the <code>-1</code> is a parameter that indicates there will be only 1 piece of content on each line. Note: <code>-2</code>, <code>-3</code> etc are not parameters that work with <code>ls</code>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> /</span></code></pre></div>
<p><strong>Reminder</strong>: you can use the <code>man</code> command to look at more options for commands.</p>
<p><code>ls</code> is my most typed command. I am consistently using it to see what directories and files are in my current directory and other directories. I suggest you do the same.</p>
</div>
<div id="mcqs-files-directories" class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> MCQs: Files &amp; Directories</h2>
<p><img src="figures/question_bubble.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Please attempt to answer the below Multiple-Choice Questions to reinforce what you have learnt in this chapter.</p>
<ol style="list-style-type: decimal">
<li>What command lists the contents of directories?
<div id="radio_ANURRQUYFT" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ANURRQUYFT" value=""></input> <span><strong><code>cp</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ANURRQUYFT" value=""></input> <span><strong><code>cd</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ANURRQUYFT" value="answer"></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What command changes directory?
<div id="radio_SPQABEJJKN" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_SPQABEJJKN" value=""></input> <span><strong><code>cp</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SPQABEJJKN" value="answer"></input> <span><strong><code>cd</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SPQABEJJKN" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What command copies files and directories?
<div id="radio_WSWPPECPNS" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_WSWPPECPNS" value="answer"></input> <span><strong><code>cp</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_WSWPPECPNS" value=""></input> <span><strong><code>cd</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_WSWPPECPNS" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>Choose the option that represents the root directory.
<div id="radio_IWWTWFCYJB" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_IWWTWFCYJB" value=""></input> <span><strong><code>~</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_IWWTWFCYJB" value="answer"></input> <span><strong><code>/</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_IWWTWFCYJB" value=""></input> <span><strong><code>..</code></strong></span></label>
</div></li>
<li>Choose the option that represents your home directory.
<div id="radio_LMAMXBRMJH" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LMAMXBRMJH" value="answer"></input> <span><strong><code>~</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LMAMXBRMJH" value=""></input> <span><strong><code>/</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LMAMXBRMJH" value=""></input> <span><strong><code>..</code></strong></span></label>
</div></li>
<li>Choose the option that represents one directory above.
<div id="radio_HVCZEUAOLA" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_HVCZEUAOLA" value=""></input> <span><strong><code>~</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_HVCZEUAOLA" value=""></input> <span><strong><code>/</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_HVCZEUAOLA" value="answer"></input> <span><strong><code>..</code></strong></span></label>
</div></li>
<li>Which path is an absolute path?
<div id="radio_LOIYXBHJTE" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LOIYXBHJTE" value=""></input> <span><strong>./Linux/</strong></span></label><label><input type="radio" autocomplete="off" name="radio_LOIYXBHJTE" value="answer"></input> <span><strong>/pub14/tea/nsc206/Linux/</strong></span></label>
</div></li>
<li>Which path is a relative path?
<div id="radio_JLMWCMNAOF" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_JLMWCMNAOF" value="answer"></input> <span><strong>./Linux</strong></span></label><label><input type="radio" autocomplete="off" name="radio_JLMWCMNAOF" value=""></input> <span><strong>/pub14/tea/nsc206/Linux/</strong></span></label>
</div></li>
</ol>

</div>
</div>
<div id="tipsandtricks" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Tips and tricks</h1>
<p><img src="figures/skateboard_trick.png" width="30%" style="display: block; margin: auto;" /></p>
<div class="webex-solution">
<button>
Chapter 5 viedo walk-through
</button>
<div class="container">
<iframe src="https://www.youtube.com/embed/968QURuAplk" frameborder="0" allowfullscreen class="video">
</iframe>
</div>
</div>
<div id="tab-complete" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Tab complete</h2>
<p>Tab complete is a method to quickly type commands and file paths without error. In bioinformatics tab complete is your best friend. The tab key looks like:
<img src="figures/tab_complete.png" width="15%" style="display: block; margin: auto;" />
Tab complete serves the following purposes:</p>
<ol style="list-style-type: decimal">
<li>To complete path names and commands</li>
<li>To list all content that starts with what you have already typed</li>
<li>To ensure you have no typos</li>
</ol>
<p>In the below examples press the tab button when you see <strong>&lt;tab&gt;</strong> and <strong>replace 2xx with your user number.</strong></p>
<p>Move into the directory <strong>“1_directory”</strong></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="bu">cd</span> /pub14/t<span class="op">&lt;</span>tab<span class="op">&gt;</span>n<span class="op">&lt;</span>tab<span class="op">&gt;</span>2xx<span class="op">&lt;</span>tab<span class="op">&gt;</span>L<span class="op">&lt;</span>tab<span class="op">&gt;</span>1_d<span class="op">&lt;</span>tab<span class="op">&gt;</span></span></code></pre></div>
<p>Print out all the content within the current directory that starts with <strong>“1_”</strong> in the file or directory name. This is carried out with a double tab.</p>
<p><strong>Note</strong>: the <strong>“./”</strong> is put before the 1 so it only looks in the current directory otherwise it will also look for commands.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">./1_</span><span class="op">&lt;</span>tab<span class="op">&gt;&lt;</span>tab<span class="op">&gt;</span></span></code></pre></div>
<p>List the contents of directory <strong>“1_1_directory”</strong></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">ls</span> 1_1<span class="op">&lt;</span>tab<span class="op">&gt;</span></span></code></pre></div>
<p>Change directory to <strong>“1_2_directory”</strong>. The <strong>“./”</strong> is not needed before the file name as tab will only look for directories because it is used for an argument of the <code>cd</code> command.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="bu">cd</span> 1_2<span class="op">&lt;</span>tab<span class="op">&gt;</span></span></code></pre></div>
<p>Change directory to <strong>example_1_part_1</strong>. The last tab will add the <strong>“/”</strong> to the end of the directory name, this informs you that you have correctly and fully typed in the directory name. However, this will not always occur if there is another directory name that starts the same but is longer. That is where double tab comes in handy.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="bu">cd</span> e<span class="op">&lt;</span>tab<span class="op">&gt;</span>1<span class="op">&lt;</span>tab<span class="op">&gt;</span>1<span class="op">&lt;</span>tab<span class="op">&gt;</span></span></code></pre></div>
<p>In this practical session I have given paths purposefully long names. This has been carried out to demonstrate the usefulness of tab complete and to encourage its use. Although they have been artificially extended in this case, in Bioinformatics long and informative path names are advised.</p>
</div>
<div id="ending-a-command" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Ending a command</h2>
<p>There are times when you will want to abandon a command on the command line. To do this simple press <strong>‘Ctrl’ + ‘c’</strong>.</p>
<p>This is useful if a command won’t respond or you noticed you have run a command with a typo or with the wrong file.</p>
</div>
<div id="history" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> History</h2>
<p><img src="figures/history.png" width="15%" style="display: block; margin: auto;" />
Linux will save commands you have previously entered. In the terminal, whilst at the command line, you can press the up and down keys to scroll through your history. You can then rerun previous commands or edit them with the left and right arrow keys, and run the edited version.</p>
</div>
<div id="clear" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Clear</h2>
<p>The <code>clear</code> command can be used to clear all the text from the terminal. This is useful for keeping a tidy terminal.</p>
</div>
<div id="bash-escape" class="section level2" number="5.5">
<h2><span class="header-section-number">5.5</span> Bash escape</h2>
<p>To continue a command on a new line on the command line use the backslash character, <strong><code>\</code></strong>. When you press <code>\</code> followed by enter, the command will not run and you will be on a new line on the command line. This can be useful for clarity and for long commands.
In the below example press “enter” after the end of a line.</p>
<p>Print to screen the term “Hello universe, today is a very nice day. Don’t you think so?”</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="bu">echo</span> <span class="dt">\</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>“Hello universe, today is a very nice day. Don’t you think so<span class="pp">?</span>”</span></code></pre></div>
<p>Notice that there is a space before the <code>\</code>. This is because there needs to be a space between the echo and text to print out. It is always recommended to use a space before a <code>\</code> to bash escape.</p>
<p>Bash escape is useful for this document as it will show if commands in this document are separate commands or one command over multiple lines.</p>
</div>
<div id="annotations" class="section level2" number="5.6">
<h2><span class="header-section-number">5.6</span> Annotations</h2>
<p><img src="figures/hashtag.png" width="15%" style="display: block; margin: auto;" /></p>
<p>You can annotate your code so it will not run. This is carried out by putting a <code>#</code> at the start of a code line.</p>
<p>An example:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="co">#This line is annotation and will not run</span></span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="co">#The below line will print out the text &quot;this line is not annotated&quot;</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&quot;this line is not annotated&quot;</span></span></code></pre></div>
<p>This is useful to give yourself information about what your code is doing and it is vital if you are creating scripts. We will also use annotations in this workbook to explain what some lines of code are doing.</p>
<p>I also find it useful to put a <code>#</code> at the front of a long command that I am typing or editing. This means the command won’t run if I accidentally press enter. Just be sure to remove the <code>#</code> at the start of the line before you want to run it.</p>
</div>
<div id="mcqs-tips-tricks" class="section level2" number="5.7">
<h2><span class="header-section-number">5.7</span> MCQs: Tips &amp; tricks</h2>
<p><img src="figures/question_bubble_blue.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Please attempt to answer the below Multiple-Choice Questions to reinforce what you have learnt in this chapter.</p>
<ol style="list-style-type: decimal">
<li>What allows you to auto fill paths and commands so you can quickly type without error?
<div id="radio_TEJFCSANPM" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_TEJFCSANPM" value=""></input> <span><strong><code>#</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_TEJFCSANPM" value="answer"></input> <span><strong>Tab complete</strong></span></label><label><input type="radio" autocomplete="off" name="radio_TEJFCSANPM" value=""></input> <span><strong><code>clear</code></strong></span></label>
</div></li>
<li>What command clears text from the terminal?
<div id="radio_QUXVBBMZOU" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_QUXVBBMZOU" value=""></input> <span><strong><code>#</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_QUXVBBMZOU" value=""></input> <span><strong>Tab complete</strong></span></label><label><input type="radio" autocomplete="off" name="radio_QUXVBBMZOU" value="answer"></input> <span><strong><code>clear</code></strong></span></label>
</div></li>
<li>What symbol is used for annotation?
<div id="radio_BEFVCVEUPN" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_BEFVCVEUPN" value="answer"></input> <span><strong><code>#</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_BEFVCVEUPN" value=""></input> <span><strong>Tab complete</strong></span></label><label><input type="radio" autocomplete="off" name="radio_BEFVCVEUPN" value=""></input> <span><strong><code>clear</code></strong></span></label>
</div></li>
<li>What keyboard short cut cancels/ends a command?
<div id="radio_YGGQMPSHYR" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_YGGQMPSHYR" value="answer"></input> <span><strong>Ctrl + c</strong></span></label><label><input type="radio" autocomplete="off" name="radio_YGGQMPSHYR" value=""></input> <span><strong><code>\</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_YGGQMPSHYR" value=""></input> <span><strong>Up and down arrow</strong></span></label>
</div></li>
<li>What keys scroll through your previously run commands (aka history)?
<div id="radio_PMBQXYQWSO" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_PMBQXYQWSO" value=""></input> <span><strong>Ctrl + c</strong></span></label><label><input type="radio" autocomplete="off" name="radio_PMBQXYQWSO" value=""></input> <span><strong><code>\</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_PMBQXYQWSO" value="answer"></input> <span><strong>Up and down arrows</strong></span></label>
</div></li>
<li>Bash escape allows you to run commands over multiple lines. What symbol is used for bash escape?
<div id="radio_AOGLBCGYHC" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_AOGLBCGYHC" value=""></input> <span><strong>Ctrl + c</strong></span></label><label><input type="radio" autocomplete="off" name="radio_AOGLBCGYHC" value="answer"></input> <span><strong><code>\</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_AOGLBCGYHC" value=""></input> <span><strong>Up and down arrow</strong></span></label>
</div></li>
</ol>

</div>
</div>
<div id="manipulatingdirectories" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> Manipulating directories</h1>
<p><img src="figures/transform.png" width="20%" style="display: block; margin: auto;" /></p>
<div class="webex-solution">
<button>
Chapter 6 video walk-through
</button>
<div class="container">
<iframe src="https://www.youtube.com/embed/5PguHkogxjI" frameborder="0" allowfullscreen class="video">
</iframe>
</div>
</div>
<div id="make-a-directory" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Make a directory</h2>
<p>To make a directory the command <code>mkdir</code> is used.</p>
<p>For the below examples we will be making heavy use of <code>ls</code> so you can hopefully visualise the contents of the directories.</p>
<p>Type in the following commands to make a subdirectory within the <strong>“CGR_Linux”</strong> directory called <strong>“Chicken”</strong> then make a subdirectory within <strong>“Chicken”</strong> called <strong>“Egg”</strong>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="co">#Change directory to CGR_linux in your home (~)</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="bu">cd</span> ~/Linux/</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="co">#Always good to list contents when you move into a directory</span></span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a><span class="co">#Make a directory called Chicken in your current directory</span></span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a><span class="fu">mkdir</span> Chicken</span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a><span class="co">#Make a subdirectory of Chicken called Egg</span></span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a><span class="fu">mkdir</span> Chicken/Egg</span>
<span id="cb26-9"><a href="#cb26-9" tabindex="-1"></a><span class="co">#List the contents of the Chicken directory</span></span>
<span id="cb26-10"><a href="#cb26-10" tabindex="-1"></a><span class="fu">ls</span> Chicken</span></code></pre></div>
<p><strong>Tip</strong>: You can use the up arrow key to get to previously run commands which you can edit using the right and left keys.</p>
</div>
<div id="moving-directories-and-files" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Moving Directories and Files</h2>
<p><img src="figures/chicken.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Files and directories can be moved with the <code>mv</code> command. This can also be used to change the name of a file or a directory.</p>
<p><strong>Note</strong>: If you move a file to the path of a file that already exists, the pre existing file will be overwritten.</p>
<p>First ensure your working directory is the <strong>“Linux”</strong> directory.</p>
<p>Move the <strong>“Chicken”</strong> directory into the <strong>“3_chicken_farm/3_1_hut/”</strong> directory. This will move the directory and all its contents.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co">#Before moving list the contents of the the destination directory</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a><span class="fu">ls</span> 3_chicken_farm/3_1_hut/</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a><span class="co">#Also list the contents of the current directory to ensure </span></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="co">#Chicken is present</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a><span class="co">#Move Chicken to 3_chicken_farm/3_1_hut/</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a><span class="fu">mv</span> Chicken/ 3_chicken_farm/3_1_hut/</span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a><span class="co">#List the current directory and destination</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a><span class="fu">ls</span> . 3_chicken_farm/3_1_hut/</span></code></pre></div>
<p>Move the Directory <strong>“3_chicken_farm/3_1_hut/Chicken/”</strong> to <strong>“3_chicken_farm/3_2_field”</strong> and rename it <strong>“Outdoor_Chicken”</strong>.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co">#List the contents of the start directory</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="fu">ls</span> ~/Linux/3_chicken_farm/3_1_hut</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="co">#Move the Chicken directory whilst renaming it</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a><span class="fu">mv</span> ~/Linux/3_chicken_farm/3_1_hut/Chicken/ <span class="dt">\</span></span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>~/Linux/3_chicken_farm/3_2_field/Outdoor_Chicken</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a><span class="co">#List the contents of the destination directory</span></span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a><span class="fu">ls</span> ~/Linux/3_chicken_farm/3_2_field/Outdoor_Chicken</span></code></pre></div>
<p>Move the file <strong>“3_chicken_farm/3_3_supplies/feed.txt”</strong> to the directory <strong>“3_chicken_farm/3_2_field”</strong> and rename it <strong>“used_feed.txt”</strong></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co">#List start directory to see if the feed.txt file exists</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="fu">ls</span> 3_chicken_farm/3_3_supplies/feed.txt</span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="co">#Move and rename the feed.txt file</span></span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a><span class="fu">mv</span> 3_chicken_farm/3_3_supplies/feed.txt <span class="dt">\</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>3_chicken_farm/3_2_field/used_feed.txt</span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a><span class="co">#List destination directory</span></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a><span class="fu">ls</span> 3_chicken_farm/3_2_field/</span></code></pre></div>
</div>
<div id="copying-directories-and-files" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Copying Directories and Files</h2>
<p><img src="figures/copy.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Files and directories can be copied with the <code>cp</code> command. Directories and files can be copied to any directory and given a new name.
Note: If you copy a file to the path of a file that already exists, the pre existing file will be overwritten.</p>
<p>Before running the below commands move into your <strong>“~/Linux/3_chicken_farm/”</strong> directory.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="co">#Change directory</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="bu">cd</span> ~/Linux/3_chicken_farm/</span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="co">#List contents of current directory</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<p>Copy the file <strong>“3_1_hut/Laid_Egg.txt”</strong> to the <strong>“3_2_field”</strong> directory and give it the filename <strong>“Chick.txt”</strong></p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co">#List contents of 3_1_hut</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="fu">ls</span> 3_1_hut</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="co">#Copy Laid_egg to another directory as a new file called Chick.txt</span></span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a><span class="fu">cp</span> 3_1_hut/Laid_Egg.txt 3_2_field/Chick.txt</span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="co">#List contents of 3_2_field</span></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a><span class="fu">ls</span> 3_2_field</span></code></pre></div>
<p>Copy the file <strong>“3_2_field/Chick.txt”</strong> and give the copy the name <strong>“Chick_2.txt”</strong></p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co">#List contents of 3_2_field</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="fu">ls</span> 3_2_field</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="co">#Copy the Chick.txt file as Chick_2.txt</span></span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a><span class="fu">cp</span> 3_2_field/Chick.txt 3_2_field/Chick_2.txt</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a><span class="co">#List contents of 3_2_field</span></span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a><span class="fu">ls</span> 3_2_field</span></code></pre></div>
<p>Copy the directory <strong>“3_1_hut”</strong> to <strong>“3_1_hut_2”</strong></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co">#List contents of current directory</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a><span class="co">#Copy directory with -r (recursively)</span></span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> 3_1_hut 3_1_hut_2</span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a><span class="co">#List contents of current directory</span></span>
<span id="cb33-6"><a href="#cb33-6" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
</div>
<div id="deleting-files-and-directories" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Deleting Files and Directories</h2>
<p><img src="figures/erase.png" width="20%" style="display: block; margin: auto;" />
To delete files and directories the command <code>rm</code> can be used.</p>
<p><strong>Important</strong>: This command can be very dangerous. There is no recycle bin on Linux machines so once you delete something it cannot be recovered. Be very careful when deleting files and directories. To avoid major loss try to keep backups of important data and scripts.</p>
<p>Ensure you are in your <strong>“3_chicken_farm/”</strong> directory</p>
<p>Remove the file <strong>“3_2_field/Chick_2.txt”</strong></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="co">#List contents of 3_2_field</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="fu">ls</span> 3_2_field</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a><span class="co">#Remove Chick_2.txt</span></span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a><span class="fu">rm</span> 3_2_field/Chick_2.txt</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a><span class="co">#List contents of 3_2_field</span></span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a><span class="fu">ls</span> 3_2_field</span></code></pre></div>
<p>Remove the directory <strong>“3_1_hut_2/”</strong> and its contents</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co">#List contents of current directory</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a><span class="co">#Remove directory with -r (recursively)</span></span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span>  3_1_hut_2</span>
<span id="cb35-5"><a href="#cb35-5" tabindex="-1"></a><span class="co">#List contents of current directory</span></span>
<span id="cb35-6"><a href="#cb35-6" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
</div>
<div id="mcqs-manipulating-directories" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> MCQs: Manipulating directories</h2>
<p><img src="figures/question_bubble_green.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Please attempt to answer the below Multiple-Choice Questions to reinforce what you have learnt in this chapter.</p>
<ol style="list-style-type: decimal">
<li>What command removes files and directories?
<div id="radio_PKAJMVERMW" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_PKAJMVERMW" value="answer"></input> <span><strong><code>rm</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_PKAJMVERMW" value=""></input> <span><strong><code>mv</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_PKAJMVERMW" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What command moves files and directories?
<div id="radio_MXURGUZDEP" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_MXURGUZDEP" value=""></input> <span><strong><code>rm</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_MXURGUZDEP" value="answer"></input> <span><strong><code>mv</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_MXURGUZDEP" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What command makes a new directory?
<div id="radio_ZKYVFJPPXL" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ZKYVFJPPXL" value=""></input> <span><strong><code>-r</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZKYVFJPPXL" value="answer"></input> <span><strong><code>mkdir</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZKYVFJPPXL" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What option is needed for moving or removing a directory?
<div id="radio_CPFYDAFBFD" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_CPFYDAFBFD" value="answer"></input> <span><strong><code>-r</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_CPFYDAFBFD" value=""></input> <span><strong><code>mkdir</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_CPFYDAFBFD" value=""></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
<li>What command list the contents of directories?
<div id="radio_GXXZMWMMZO" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_GXXZMWMMZO" value=""></input> <span><strong><code>-r</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_GXXZMWMMZO" value=""></input> <span><strong><code>mkdir</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_GXXZMWMMZO" value="answer"></input> <span><strong><code>ls</code></strong></span></label>
</div></li>
</ol>

</div>
</div>
<div id="exercise-1" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Exercise 1</h1>
<p><img src="figures/exercise.png" width="20%" style="display: block; margin: auto;" />
Perform the following tasks with the skills and knowledge you have gained.</p>
<p>Change the name of the subdirectory <strong>“four_exercises”</strong> within your <strong>“Linux”</strong> directory to <strong>“4_exercises”</strong></p>
<ol style="list-style-type: decimal">
<li>Make a backup of the <strong>“4_exercises”</strong> directory</li>
<li>List the contents of the <strong>“4_exercises”</strong> directory</li>
<li>Within the directory <strong>“4_exercises”</strong>
<ol style="list-style-type: decimal">
<li>Print the working directory.</li>
<li>Print out to screen the phrase ‘<strong>the echo command allows me to print phrases to screen</strong>’.</li>
<li>Copy the file <strong>“copy_this_file.txt”</strong> to the directory <strong>“to_me”</strong></li>
<li>Rename the directory <strong>“to_me”</strong> to <strong>“you”</strong></li>
<li>Delete the initial <strong>“copy_this_file.txt”</strong> file</li>
</ol></li>
</ol>
<p>You can check my solutions in the <a href="#exercise1_answers">Answers section</a>. These are not the definitive solution but only examples of solutions. If your method works and you understand why then you have done it correctly.</p>

</div>



<div id="filereadingandprocessing" class="section level1" number="8">
<h1><span class="header-section-number">Chapter 8</span> File reading and processing</h1>
<p><img src="figures/files.png" width="20%" style="display: block; margin: auto;" />
There are many ways to show the contents of a file. Below are a few examples.</p>
<p>The files for the examples are within the directory: <strong>“/pub14/tea/nsc2xx/Linux/5_reading_files/”</strong> <strong>(replace xxx with your user number)</strong>.</p>
<div class="webex-solution">
<button>
Chapter 8 video walk-through
</button>
<div class="container">
<iframe src="https://www.youtube.com/embed/rldLi1yAE58" frameborder="0" allowfullscreen class="video">
</iframe>
</div>
</div>
<div id="print-out-a-file" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Print out a file</h2>
<p><img src="figures/cat.png" width="20%" style="display: block; margin: auto;" />
The <code>cat</code> command will print out the entire contents of a file to the screen. This is useful for small text files and pipelines (pipelines are not covered here).
Example commands are below (remember to replace xxx with your user name number):</p>
<p><strong>Note</strong>: remember tab complete and using the arrow keys</p>
<p>Print contents of <strong>“short_file.txt”</strong> to screen</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="fu">cat</span> /pub14/tea/nsc2xx/Linux/5_reading_files/short_file.txt</span></code></pre></div>
<p>Print contents of <strong>“Scientist.txt”</strong> to screen</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">cat</span> /pub14/tea/nsc2xx/Linux/5_reading_files/Scientist.txt</span></code></pre></div>
<p>Print contents of <strong>“ecoli.gbk”</strong> to screen</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="fu">cat</span> /pub14/tea/nsc2xx/Linux/5_reading_files/ecoli.gbk</span></code></pre></div>
<p><strong>Remember</strong>: The <code>clear</code> command.</p>
</div>
<div id="head-and-tail" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> head and tail</h2>
<p><img src="figures/head_and_tails.png" width="30%" style="display: block; margin: auto;" />
The <code>head</code> command will print out to screen the top n lines of a file.</p>
<p>The <code>tail</code> command will print out to screen the bottom n lines of a file.</p>
<p>The default value is 10. The <code>-n</code> option can be used to indicate how many lines to print out.</p>
<p>Carry out the below commands in the directory <strong>“/pub14/tea/nsc2xx/Linux/5_reading_files/”</strong></p>
<p>Print out the top 10 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb39"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">head</span> ecoli.gbk</span></code></pre></div>
<p>Print out the bottom 10 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="fu">tail</span> ecoli.gbk</span></code></pre></div>
<p>Print out the top 25 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 25 ecoli.gbk</span></code></pre></div>
<p>Print out the bottom 2 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> 2 ecoli.gbk</span></code></pre></div>
<p>Print out all but the bottom 2 lines of <strong>“Scientist.txt”</strong></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> <span class="at">-2</span> Scientist.txt</span></code></pre></div>
<p>Print out all lines starting from the 2nd top line of <strong>“Scientist.txt”</strong></p>
<div class="sourceCode" id="cb44"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> +2 Scientist.txt</span></code></pre></div>
<p>Print out all but the bottom 5 lines of <strong>“Scientist.txt”</strong></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> <span class="at">-5</span> Scientist.txt</span></code></pre></div>
<p>Print out all lines starting from the 3rd top line of <strong>“Scientist.txt”</strong></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> +3 Scientist.txt</span></code></pre></div>
<p>Print out the top 25 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> +25 ecoli.gbk</span></code></pre></div>
<p>Print out the bottom 2 lines of <strong>“ecoli.gbk”</strong></p>
<div class="sourceCode" id="cb48"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> <span class="at">-2</span> ecoli.gbk</span></code></pre></div>
</div>
<div id="file-viewing-with-less" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> File viewing with less</h2>
<p><img src="figures/less.png" width="20%" style="display: block; margin: auto;" />
The <code>less</code> command will display a file’s contents one page at a time. Various keys on the keyboard will allow you to navigate the contents of the files. The below actions will occur identically with the man command.</p>
<ul>
<li><strong>q</strong> : Exit</li>
<li><strong>up and down arrow keys</strong> : Will move up/down 1 line at a time</li>
<li><strong>space</strong> : Move down one page</li>
<li><strong>b</strong> : Move up one page</li>
<li><strong><code>/</code></strong> : Follow this by a term to search for it in the file’s contents</li>
<li><strong>n</strong> : Find the next occurrence of the term last searched for</li>
<li><strong>N</strong> : Find the previous occurrence of the term last searched for</li>
<li><strong>g</strong> : Jump to the first line of the file</li>
<li><strong>G</strong> : Jump to the bottom line of the file</li>
</ul>
<p>Use the <code>less</code> command to view the contents of the <strong>“ecoli.gbk”</strong> file. Then find the 3rd occurrence of the word ‘ribosome’. Afterwards move around the file.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">less</span> ecoli.gbk</span></code></pre></div>
<p>Look at the <code>man</code>ual for <code>less</code> and search for the first occurrence of the string ‘percent’. Afterwards look around the manual page.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">man</span> less</span></code></pre></div>
</div>
<div id="word-count" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Word count</h2>
<p><img src="figures/wc.png" width="20%" style="display: block; margin: auto;" />
The <code>wc</code> command will allow allow you to word count files. It will display line, word and byte counts for files in that order.</p>
<p>Use <code>wc</code> to see the line, word and byte count of the <strong>“short_file.txt”</strong> and <strong>“ecoli.gbk”</strong> files. As you can see you can carry this out on multiple files at once.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">wc</span> short_file.txt Scientist.txt ecoli.gbk</span></code></pre></div>
<p>Count the number of characters in the <strong>“short_file.txt”</strong> file</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-m</span> short_file.txt</span></code></pre></div>
<p>Count the number of lines in the <strong>“ecoli.gbk”</strong> file</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> ecoli.gbk</span></code></pre></div>
</div>
<div id="pattern-searching" class="section level2" number="8.5">
<h2><span class="header-section-number">8.5</span> Pattern searching</h2>
<p><img src="figures/pattern.png" width="20%" style="display: block; margin: auto;" />
The <code>grep</code> command will search for a pattern in a text file and output all the lines containing the pattern.</p>
<p>Print out the lines from <strong>“Scientist.txt”</strong> that have the number 18 in them. In this particular example it prints out all scientists which were born in the 1800s. This will not always be the case depending on the data in the file.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="fu">grep</span> “18” Scientist.txt</span></code></pre></div>
<p>Print out the lines which have the string “Ada” in them.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="fu">grep</span> “Ada” Scientist.txt</span></code></pre></div>
<p>Print out the lines which have the string “ada” in them. There should be none, as grep is case sensitive.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">grep</span> “ada” Scientist.txt</span></code></pre></div>
<p>Type in the following command.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="fu">grep</span> Scientist.txt</span></code></pre></div>
<p>The above command will be stuck as <code>grep</code> does not know what it is looking for. To cancel the command use <strong>‘Ctrl’ + ‘c’</strong></p>
</div>
<div id="text-editor" class="section level2" number="8.6">
<h2><span class="header-section-number">8.6</span> Text editor</h2>
<p><img src="figures/text_editor.png" width="20%" style="display: block; margin: auto;" />
Three of the most popular text editors are <strong>vim</strong>, <strong>gedit</strong> and <strong>nano</strong>. Below are quick introductions to <strong>nano</strong> and <strong>vim</strong>.</p>
<p><strong>nano</strong> is the easiest to learn but is quite limiting. <strong>vim</strong> and <strong>gedit</strong> are quite similar in power with different people preferring one or the other.</p>
<p>The below will teach you <strong>nano</strong>. If you are interested in learning <strong>vim</strong> in the future you can find a quick guide in the <a href="#vim">appendix</a>.</p>
<div id="nano" class="section level3" number="8.6.1">
<h3><span class="header-section-number">8.6.1</span> nano</h3>
<p><img src="figures/1200px-Gnu-nano.png" width="20%" style="display: block; margin: auto;" />
To enter the <code>nano</code> text editor you can use the command <code>nano</code>. The command is: <code>nano file.txt</code>.</p>
<p><code>nano</code> can be run with a previous file name which you can then edit or a new file name in which case you will create a new file.</p>
<p>Once you are in the editor you can type characters and move around with the arrow keys.</p>
<p>To carry out specific functions you will need to use <strong>Ctrl</strong> or <strong>Alt</strong> with another key. At the bottom of the editor are a few examples where the <code>^</code> indicates <strong>Ctrl</strong>. For example the <code>^G Get Help</code> means you need to press <strong>Ctrl+G</strong> to get help. When you use letters this way in <strong>nano</strong> they are case insensitive (i.e. the CAPS lock can be on or off and you will get the same result).</p>
<p>After you carry out a function ensure you look at the bottom of the editor again as it may ask you to type something or you may get a new series of functions you can use.</p>
<p>Below are some important examples:</p>
<ul>
<li><strong>Ctrl+X</strong> - Exit nano</li>
<li><strong>Ctrl+S</strong> - Save file</li>
<li><strong>Ctrl+O</strong> - Save file as</li>
<li><strong>Ctrl+A</strong> - Jump to the start of a line</li>
<li><strong>Ctrl+E</strong> - Jump to the end of a line</li>
<li><strong>Ctrl+W</strong> - Start search (Where is) <strong>Note</strong> This unfortunately is also the shortcut to close a tab in internet browsers. Therefore this can’t be used with our webVNC.</li>
<li><strong>Alt+W</strong> - Continue search forward (find next occurrence forward)</li>
<li><strong>Alt+Q</strong> - Find next occurrence backward</li>
<li><strong>Alt+K</strong> - Cut current line</li>
<li><strong>Alt+\</strong> - Go to the first line</li>
<li><strong>Alt+/</strong> - Go to the last line</li>
</ul>
<p><a href="https://www.nano-editor.org/dist/latest/cheatsheet.html"><strong>Nano cheatsheet</strong></a></p>
</div>
<div id="tasks" class="section level3" number="8.6.2">
<h3><span class="header-section-number">8.6.2</span> Tasks</h3>
<p>Carry out the following tasks in the directory: <strong>“/pub14/tea/nscxx/Linux/5_reading_files/”</strong></p>
<p>Using a text editor (<strong>nano</strong> or <strong>vim</strong>) add an entry for Scientist Mae Jemison (Born: 1956) to the file <strong>“Scientist.txt”</strong>. The names and date are separated by one tab.</p>
<p>Using your text editor of choice delete all the scientists born before 1000 in the <strong>“Scientist.txt”</strong> file and save this as <strong>“Scientist_post_1000.txt”</strong>.</p>
</div>
</div>
<div id="mcqs-file-reading-and-processing" class="section level2" number="8.7">
<h2><span class="header-section-number">8.7</span> MCQs: File reading and processing</h2>
<p><img src="figures/question_bubble_purple.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Please attempt to answer the below Multiple-Choice Questions to reinforce what you have learnt in this chapter.</p>
<ol style="list-style-type: decimal">
<li>What command searches for a pattern?
<div id="radio_LARPQWIOPU" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LARPQWIOPU" value=""></input> <span><strong><code>wc</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LARPQWIOPU" value="answer"></input> <span><strong><code>grep</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LARPQWIOPU" value=""></input> <span><strong><code>cat</code></strong></span></label>
</div></li>
<li>What command word counts files?
<div id="radio_SEPJPEVMET" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_SEPJPEVMET" value="answer"></input> <span><strong><code>wc</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SEPJPEVMET" value=""></input> <span><strong><code>grep</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SEPJPEVMET" value=""></input> <span><strong><code>cat</code></strong></span></label>
</div></li>
<li>What command prints the contents of a file?
<div id="radio_ZLUQFHGQRO" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ZLUQFHGQRO" value=""></input> <span><strong><code>wc</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZLUQFHGQRO" value=""></input> <span><strong><code>grep</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZLUQFHGQRO" value="answer"></input> <span><strong><code>cat</code></strong></span></label>
</div></li>
<li>What command displays a file’s contents one page at a time and allows keyboard navigation?
<div id="radio_FJUQLCKZXT" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_FJUQLCKZXT" value="answer"></input> <span><strong><code>less</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_FJUQLCKZXT" value=""></input> <span><strong><code>tail</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_FJUQLCKZXT" value=""></input> <span><strong><code>head</code></strong></span></label>
</div></li>
<li>What command prints out the top n lines of a file
<div id="radio_YMXXEUIEOW" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_YMXXEUIEOW" value=""></input> <span><strong><code>less</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_YMXXEUIEOW" value=""></input> <span><strong><code>tail</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_YMXXEUIEOW" value="answer"></input> <span><strong><code>head</code></strong></span></label>
</div></li>
<li>What command prints out the bottom n lines of a file
<div id="radio_XPZCPSQFON" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_XPZCPSQFON" value=""></input> <span><strong><code>less</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_XPZCPSQFON" value="answer"></input> <span><strong><code>tail</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_XPZCPSQFON" value=""></input> <span><strong><code>head</code></strong></span></label>
</div></li>
</ol>

</div>
</div>
<div id="recap" class="section level1" number="9">
<h1><span class="header-section-number">Chapter 9</span> Recap</h1>
<p><img src="figures/recap.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Throughout this practical we have covered the topics:</p>
<ul>
<li>Linux
<ul>
<li>Multiuser Multitasking OS</li>
<li>User, Shell and Kernel</li>
</ul></li>
<li>Running commands on the command line
<ul>
<li>Search engines, cheat sheets and manual pages</li>
<li>Commands: <code>man</code></li>
</ul></li>
<li>Useful Linux practice and commands
<ul>
<li>Tab complete</li>
<li>Up and Down arrow keys to access history</li>
<li><strong>‘Ctrl’ + ‘c’</strong> to stop a command</li>
<li>Commands: <code>clear</code></li>
</ul></li>
<li>Files and Directories
<ul>
<li>Navigating through directory structure and looking at files with paths</li>
<li>Commands: <code>cd</code>, <code>pwd</code>, <code>ls</code>, <code>mkdir</code></li>
</ul></li>
<li>Manipulating files i.e. creating, copying, moving, deleting
<ul>
<li>Commands: <code>cp</code>, <code>mv</code>, <code>rm</code></li>
</ul></li>
<li>Looking at files
<ul>
<li>Commands: <code>cat</code>, <code>head</code>, <code>tail</code>, <code>less</code>, <code>wc</code>, <code>grep</code></li>
</ul></li>
<li>Text editor
<ul>
<li>Commands: <code>vim</code></li>
</ul></li>
</ul>
<p><strong>Note</strong>: There is a cheatsheet at the end of this book with these commands.</p>

</div>
<div id="fastq-format" class="section level1" number="10">
<h1><span class="header-section-number">Chapter 10</span> Fastq format</h1>
<p><img src="figures/files.png" width="20%" style="display: block; margin: auto;" />
The next exercise will focus on a set of files including fastq files.</p>
<ul>
<li>Fastq files are very commonly used in bioinformatics.</li>
<li>Fastq files contain DNA or Amino acid sequenceing data.</li>
<li>Fastq files contain the nucleotide/amino acid content and its sequencing quality for sequences.</li>
<li>Generally these files are separated by sample but not always.</li>
<li>A fastq file acts as a normal txt file that can be read but is of a specific format.</li>
<li>One fastq file contains many fastq entries, one after the other.</li>
<li>Each fastq entry contains four lines.
<ul>
<li>One fastq entry represents one sequence.</li>
</ul></li>
</ul>
<p>The format of one entry is as below:</p>
<p><strong>@Sequence 1</strong><br />
<strong>CTGTTAAATACCGACTTGCGTCAGGTGCGTGAACAACTGGGCCGCTTT</strong><br />
<strong>+</strong><br />
<strong>=&lt;&lt;&lt;=&gt;@@@ACDCBCDAC@BAA@BA@BBCBBDA@BB@&gt;CD@A@B?B@@</strong></p>
<p>The lines represent:<br />
1. Header for fastq entry known as the fastq header. This always begins with a ‘@’.<br />
2. Sequence content of sequence<br />
3. Quality header. Always begins with a ‘+’. Sometimes also contains the same information as fastq header.<br />
4. Quality values for each base in the 2nd line. NOTE: ‘@’ can be used as quality values.</p>
<p>For more information on the fastq format the below resource is good:
<a href="https://en.wikipedia.org/wiki/FASTQ_format" class="uri">https://en.wikipedia.org/wiki/FASTQ_format</a></p>

</div>
<div id="exercise-2" class="section level1" number="11">
<h1><span class="header-section-number">Chapter 11</span> Exercise 2</h1>
<p><img src="figures/exercise_2.png" width="15%" style="display: block; margin: auto;" />
The directory <strong>“~/Linux/6_final_exercise/”</strong> has all the files you need. Below is a set of tasks and questions that will require all the skills you have gained from this practical.</p>
<ol style="list-style-type: decimal">
<li>See what files are in the directory.<br />
</li>
<li>Rename the file “3-P£_CACTTCGA_L001_R1_001.fastq” as “3-P3_CACTTCGA_L001_R1_001.fastq”.<br />
</li>
<li>Make a backup of the files in a directory called <strong>backup</strong>.<br />
</li>
<li>How many reads are in the samples?<br />
</li>
<li>Remove the fastq files with no data.</li>
<li>Update the backup files with the previous change.<br />
</li>
<li>Check if the 1st read names match in the paired files.<br />
</li>
<li>Check if the last read names match in the paired files.<br />
</li>
<li>In file <strong>“1-P1_ATGCCTGG_L001_R1_001.fastq”</strong> look for sequence headers with the term ‘psychrobacter’.<br />
</li>
<li>In the sample <strong>1-P1</strong> remove any fastq entries where the term ‘psychrobacter’ appears in the fastq header. Do this for the R1 and R2 files.<br />
</li>
<li>Print to screen the fastq header, sequence and quality data for the 25th sequence in sample <strong>2-P2</strong> for both the R1 and R2 file. Do this with one command for R1 and a separate command for R2.</li>
</ol>
<p>You can check my solutions in the <a href="#exercise2_answers">Answers section</a>. These are not the definitive solution but only examples of solutions. If your method works and you understand why then you have done it correctly.</p>

</div>



<div id="advancedlinux" class="section level1" number="12">
<h1><span class="header-section-number">Chapter 12</span> Advanced Linux practice</h1>
<p><img src="figures/linux_intermdiary.png" width="20%" style="display: block; margin: auto;" />
We have covered a small amount of Linux coding. This should be sufficient to carry out our future workshops but if you were to continue in bioinformatics we would recommend learning more advanced methods.</p>
<p>Below are some short sections to introduce you to some more advanced linux coding techniques. These give you a quick overview and some examples. This will hopefully put you in a good position to allow you to to learn these techniques in more depth outside of this workshop.</p>
<p>The following sections will all be run with the files in the directory <strong>“~/Linux/advanced_practice/”</strong>. Therefore ensure you are in this directory before running the below examples. This contains fastq and txt files for 20 samples. Each sample contains a fastq file and txt file for uncorrected and corrected reads. These fastq files are single end (i.e. there is no reverse/R2 reads).</p>
<div id="wildcard-characters" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> Wildcard characters</h2>
<p><img src="figures/wildcard.png" width="20%" style="display: block; margin: auto;" />
These are characters that can be used to represent a variety of other characters. This can be useful for deleting many files, searching for files with specific patterns in their names and more!</p>
<p>Be very careful when using wildcard character with the command <strong>rm</strong>.</p>
<p>Three basic and useful wildcards include:</p>
<ul>
<li><code>*</code> - This represents zero or more characters<br />
</li>
<li><code>?</code> - This represents a single character<br />
</li>
<li><code>[]</code> - This represents a range of characters</li>
</ul>
<p>Below are various examples you can run to show wildcards in action.</p>
<p>List all the files and directories in the working directory</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span></span></code></pre></div>
<p>List all files ending in “.fastq”</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>.fastq</span></code></pre></div>
<p>List all files ending in “.txt”</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>.txt</span></code></pre></div>
<p>List all files with the string “corrected” somewhere in the file name</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>corrected<span class="pp">*</span></span></code></pre></div>
<p>List all files with the string “corrected” somewhere in the file name and it ends in “fastq”</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>corrected<span class="pp">*</span>fastq</span></code></pre></div>
<p>List all files that begin with “sample_2”</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="fu">ls</span> sample_2<span class="pp">*</span></span></code></pre></div>
<p>List all files that begin with “sample_2_”</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a><span class="fu">ls</span> sample_2_<span class="pp">*</span></span></code></pre></div>
<p>List the files that begin with “sample_1” and ends with “AAAA.txt”. It may have zero or more characters between these two.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="fu">ls</span> sample_1_<span class="pp">*</span>AAAA.txt</span></code></pre></div>
<p>List the fastq files of samples with a single digit number</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a><span class="fu">ls</span> sample_<span class="pp">?</span>_<span class="pp">*</span>fastq</span></code></pre></div>
<p>List the txt files of samples with a number in the tens</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="fu">ls</span> sample_1<span class="pp">?</span>_<span class="pp">*</span>txt</span></code></pre></div>
<p>List the txt files for samples 3,4,5,6 &amp; 7 i.e. 3-7</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a><span class="fu">ls</span> sample_<span class="pp">[</span><span class="ss">3</span><span class="pp">-</span><span class="ss">7</span><span class="pp">]</span>_<span class="pp">*</span>txt</span></code></pre></div>
<p>List the txt files for the non corrected information of samples with single digits.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="fu">ls</span> sample_<span class="pp">?</span>_<span class="pp">*[</span><span class="ss">ATGC</span><span class="pp">]</span>.txt</span></code></pre></div>
<p>List the corrected txt files for samples with numbers divisible by 10.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a><span class="fu">ls</span> sample_<span class="pp">[</span><span class="ss">0</span><span class="pp">-</span><span class="ss">9</span><span class="pp">][</span><span class="ss">0</span><span class="pp">]</span>_<span class="pp">*</span>corrected.txt</span></code></pre></div>
</div>
<div id="redirection" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> Redirection</h2>
<p><img src="figures/chevron.png" width="15%" style="display: block; margin: auto;" />
Redirection allows you to put the output of a command to a file. The redirect symbol is <code>&gt;</code>. Be careful when redirecting as it will overwrite any existing files. To append to the bottom of a file use <code>&gt;&gt;</code>.</p>
<p>Below are various examples of redirecting in action.</p>
<p>Create a file called <strong>ecoli.tmp</strong> containing the text “I am escherichia coli”</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="bu">echo</span> “I am escherichia coli” <span class="op">&gt;</span> ecoli.tmp</span></code></pre></div>
<p>Create a file called <strong>pcryohalolentis.tmp</strong> containing the text “I am psychrobacter cryohalolentis”</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a><span class="bu">echo</span> “I am psychrobacter cryohaloloentis” <span class="op">&gt;</span> pcryohalolentis.tmp</span></code></pre></div>
<p>Create a new file called <strong>bacteria.tmp</strong> which will contain the text from <strong>ecoli.tmp</strong> and <strong>pcryohalolentis.tmp</strong></p>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb73-1"><a href="#cb73-1" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>tmp <span class="op">&gt;</span> bacteria.tmp</span></code></pre></div>
<p>Create a file called <strong>vcholerae.tmp</strong> containing the text “I am not ecoli or pcryohalolentis”</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="bu">echo</span> “I am not ecoli or pcryohalolentis” <span class="op">&gt;</span> vcholerae.tmp</span></code></pre></div>
<p><code>cat</code> the file <strong>vcholerae.tmp</strong> and redirect it to <strong>bacteria.tmp</strong>.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a><span class="fu">cat</span> vcholerae.tmp <span class="op">&gt;</span> bacteria.tmp</span></code></pre></div>
<p>Look at the contents of <strong>bacteria.tmp</strong></p>
<div class="sourceCode" id="cb76"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="fu">cat</span> bacteria.tmp</span></code></pre></div>
<p>This has removed the ecoli and pcryohalolentis lines. Append the contents of <strong>ecoli.tmp</strong> and <strong>pcryohalolentis.tmp</strong> to <strong>bacteria.tmp</strong> and then check the file</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="fu">cat</span> ecoli.tmp pcryohalolentis.tmp <span class="op">&gt;&gt;</span> bacteria.tmp</span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a><span class="fu">cat</span> bacteria.tmp</span></code></pre></div>
<p>Put information regarding number of lines of all the fastq files into a new file called
<strong>fastq_lines.tmp</strong></p>
<div class="sourceCode" id="cb78"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> <span class="pp">*</span>fastq <span class="op">&gt;</span> fastq_lines.tmp</span></code></pre></div>
<p>Now delete all the files that were created in the above examples. Again be very careful about using the <code>rm</code> command with wildcards.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a><span class="fu">rm</span> <span class="pp">*</span>tmp</span></code></pre></div>
</div>
<div id="pipes" class="section level2" number="12.3">
<h2><span class="header-section-number">12.3</span> Pipes</h2>
<p><img src="figures/pipes.png" width="20%" style="display: block; margin: auto;" />
Pipes allow you to put the output of one command to the input of another. For example you could use <code>grep</code> to get all the lines with a certain string and pipe the output to <code>wc</code> to count the number of lines that have the specific string.</p>
<p>The pipe symbol is <code>|</code>. This is normally found on your keyboard directly left of the Z key. Weirdly the symbol is represented by <code>|</code> but split in the middle on some keyboards.</p>
<p>A useful tip when building up longer pipes is to start with a smaller amount of data and check the output of each step as you go. To do this you could use <code>head</code> instead of <code>cat</code> whilst testing.</p>
<p>Below are various examples of piping in action</p>
<p>Print to screen the second last fastq entry of the file <strong>sample_20_ATAC_corrected.fastq</strong></p>
<div class="sourceCode" id="cb80"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a><span class="fu">cat</span> sample_20_ATAC_corrected.fastq <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> 8 <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 4 </span></code></pre></div>
<p><strong>Note</strong>: In the above command the <code>tail</code> command is working on the output of the <code>cat</code> command. Therefore this would not work to get the second last fastq entry of multiple files. For example the following command would print the second last fastq entry of the last fastq file (i.e. <strong>sample_9_AAGA.fastq</strong> due to file ordering)</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>fastq <span class="kw">|</span> <span class="fu">tail</span> <span class="at">-n</span> 8 <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 4</span></code></pre></div>
<p>Count the number of lines within all the fastq files</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>fastq <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
<p>Count the number of lines which contain the text “TAG” within all the fastq files</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>fastq <span class="kw">|</span> <span class="fu">grep</span> “TAG” <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre></div>
</div>
<div id="regular-expressions" class="section level2" number="12.4">
<h2><span class="header-section-number">12.4</span> Regular expressions</h2>
<p><img src="figures/trex.png" width="20%" style="display: block; margin: auto;" />
Regular expressions are similar to wildcard characters but more complex and used for commands like <code>grep</code> and <code>sed</code>.</p>
<p>Below are a basic set of regular expressions:</p>
<ul>
<li><code>.</code> : A single character<br />
</li>
<li><code>?</code> : The preceding character matches 1 or 0 times<br />
</li>
<li><code>*</code> : The preceding character matches zero or more times<br />
</li>
<li><code>+</code> : The preceding character matches one or more times<br />
</li>
<li><code>{n}</code> : The preceding character matches exactly n times<br />
</li>
<li><code>{n,m}</code> : The preceding character matches n to m times<br />
</li>
<li><code>[AT]</code> : The character is one of the characters in the brackets<br />
</li>
<li><code>[^CG]</code> : The character is not one of those in the brackets<br />
</li>
<li><code>[1-7]</code> : The character is 1,2,3,4,5,6 or 7. This works with letters too.<br />
</li>
<li><code>()</code> : Group several characters into one<br />
</li>
<li><code>|</code> : Logical OR operator<br />
</li>
<li><code>^</code> : Matches the beginning of the line<br />
</li>
<li><code>$</code> : Matches the end of the line</li>
</ul>
<p>Below are various examples of regular expressions in action.</p>
<p>Look at the contents of <strong>metadata.txt</strong></p>
<div class="sourceCode" id="cb84"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt</span></code></pre></div>
<p>!Print out the lines for the Healthy patients</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb85-1"><a href="#cb85-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">grep</span> “HEALTHY”</span></code></pre></div>
<p>Print out the lines for the IBD patients from Craigavon and Belfast. In the below command <code>\</code> is used to allow <code>|</code> to be used as an <strong>or</strong> operator instead of acting as a string to match.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;IBD&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;BELFAST\|CRAIGAVON&quot;</span></span></code></pre></div>
<p>Print out the lines for the Pre information of patients not from Edinburgh or Aberdeen</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb87-1"><a href="#cb87-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;PRE$&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;ABERDEEN\|EDINBURGH&quot;</span></span></code></pre></div>
<p>Print out the lines for patients 1,2,3 and 4</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;Patient_[1-4][^0-9]&quot;</span> </span></code></pre></div>
<p>Print to screen every fastq header of file <strong>sample_15_AACG_corrected.fastq</strong></p>
<div class="sourceCode" id="cb89"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb89-1"><a href="#cb89-1" tabindex="-1"></a><span class="fu">cat</span> sample_15_AACG_corrected.fastq <span class="kw">|</span> <span class="fu">grep</span> “^@sample”</span></code></pre></div>
<p>In the piping examples we counted the number of lines with the text “TAG” within the fastq files. However this also counted fastq headers due to the name of the samples. Let us use a regular expression to only count the number of sequences within the fastq files that contain “TAG”.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="fu">cat</span> <span class="pp">*</span>fastq <span class="kw">|</span> <span class="fu">grep</span> “^<span class="pp">[^</span><span class="ss">@</span><span class="pp">]</span>.<span class="pp">*</span>TAG”</span></code></pre></div>
<p>Print to screen every line within the file <strong>sample_3_AAAG_corrected.fastq</strong> that has a possible Threonine codon in the forward direction.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb91-1"><a href="#cb91-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">&quot;^[^@].*AC[ACTG]&quot;</span> sample_3_AAAG_corrected.fastq</span></code></pre></div>
<p>In the above example fastq quality lines are also extracted as some of them also contain the pattern we are searching for. To get around this we can pipe. First grep the fastq quality header (i.e. +), as no other line only contains “+”, and the line before it. Then we can remove lines with only a plus with an invert grep. Finally we can grep for the threonine pattern using only the sequence lines. Let us build this up step by step.</p>
<p>Print to screen the fastq quality header plus the one line preceding each (i.e. Sequence line) for file <strong>sample_3_AAAG_corrected.fastq.</strong></p>
<div class="sourceCode" id="cb92"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B</span> 1 “^+$” sample_3_AAAG_corrected.fastq</span></code></pre></div>
<p>Now pipe this output so it removes the lines with “+” (fastq quality headers) and “–” (separators of each <code>grep</code> match provided by grep because of the -B 1 option).</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B</span> 1 <span class="st">&quot;^+$&quot;</span> sample_3_AAAG_corrected.fastq <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;+\|--&quot;</span></span></code></pre></div>
<p>Now from this output, <code>grep</code> for the Threonine pattern plus colour each match within the line with the option “<strong>–color</strong>”.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B</span> 1 <span class="st">&quot;^+$&quot;</span> sample_3_AAAG_corrected.fastq <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;+\|--&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">--color</span> <span class="st">&quot;AC[ACTG]&quot;</span></span></code></pre></div>
<p>Let us repeat the above but add the possibility of the threonine being in the reverse direction.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb95-1"><a href="#cb95-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B</span> 1 <span class="st">&quot;^+$&quot;</span> sample_3_AAAG_corrected.fastq <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;+\|--&quot;</span> <span class="dt">\</span></span>
<span id="cb95-2"><a href="#cb95-2" tabindex="-1"></a><span class="kw">|</span> <span class="fu">grep</span> <span class="at">--color</span> <span class="st">&quot;AC[ACTG]\|[ACTG]CA&quot;</span></span></code></pre></div>
<p><strong>Resources to learn more in the future</strong></p>
<p>Rex Egg, A good resource to learn more about regular expressions:</p>
<p><a href="https://www.rexegg.com/" class="uri">https://www.rexegg.com/</a></p>
<p>Cheatsheet:</p>
<p><a href="https://www.rexegg.com/regex-quickstart.html" class="uri">https://www.rexegg.com/regex-quickstart.html</a></p>
<p>Regex Crossword, A online game like soduku that is useful to practice regular expressions. Best used in conjunction with the above cheat sheet:</p>
<p><a href="https://regexcrossword.com/" class="uri">https://regexcrossword.com/</a></p>
</div>
<div id="sed" class="section level2" number="12.5">
<h2><span class="header-section-number">12.5</span> sed</h2>
<p><img src="figures/sed.png" width="20%" style="display: block; margin: auto;" />
This is a complicated yet powerful command that can be used to edit text files quickly and efficiently. The main use is to substitute text with other text. <code>sed</code> can be used with regular expressions.</p>
<p>The basic outline of a <code>sed</code> substitute command is as below. In the below case <code>s/</code> signifies that <code>sed</code> will be used for substitution</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="fu">sed</span> “s/old_text/new_text/” old_file <span class="op">&gt;</span> new_file</span></code></pre></div>
<p>Below are some examples of sed in action.</p>
<p>Print out a list of all the sample names using the fastq files</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb97-1"><a href="#cb97-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> <span class="pp">*[</span><span class="ss">AGTC</span><span class="pp">]</span>.fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/.fastq//&quot;</span></span></code></pre></div>
<p>First print out the contents of the file <strong>metadata.txt</strong></p>
<div class="sourceCode" id="cb98"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb98-1"><a href="#cb98-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt</span></code></pre></div>
<p>Print out <strong>metadata.txt</strong> and change IBD to DISEASE without altering the file</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb99-1"><a href="#cb99-1" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&quot;s/IBD/DISEASE/&quot;</span> metadata.txt</span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb100-1"><a href="#cb100-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/IBD/DISEASE/&quot;</span></span></code></pre></div>
<p><code>sed</code> is case sensitive and will by default only replace the first instance it finds within each line.
Print <strong>metadata.txt</strong> to screen and then change the “P” in “Patient” to “Human_P”</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb101-1"><a href="#cb101-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/P/Human_P/&quot;</span></span></code></pre></div>
<p>To replace every instance of the old pattern within each line <code>g</code> can be added after the last <code>/</code>. This stands for global therefore it changes the command to a global substitute.</p>
<p>Print <strong>metadata.txt</strong> to screen and change every occurrence of a number to “number”. To get the regular expression meaning of “+” it needs a “<strong><code>\</code></strong>” before the “+”</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb102-1"><a href="#cb102-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/[0-9]\+/number/g&quot;</span></span></code></pre></div>
<p>The file metadata.txt is tab delimited (i.e. there is a tab between each column. Make a comma separated file containing the information of <strong>metadata.txt</strong> called <strong>metadata.csv</strong> (csv = comma separated value). <code>\t</code> presents a tab.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb103-1"><a href="#cb103-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/\t/,/g&quot;</span> <span class="op">&gt;</span> metadata.csv</span></code></pre></div>
<p>For a very in depth look into the sed command please look at the following link:
<a href="http://www.grymoire.com/Unix/Sed.html" class="uri">http://www.grymoire.com/Unix/Sed.html</a></p>
</div>
<div id="permissions" class="section level2" number="12.6">
<h2><span class="header-section-number">12.6</span> Permissions</h2>
<p>All files, directories and programs have permissions. It is important to know about this so you know your read, write and executability permissions for the content within machines.</p>
<p>Below is a useful link to learn about file permissions:<br />
<a href="https://www.guru99.com/file-permissions.html" class="uri">https://www.guru99.com/file-permissions.html</a></p>
</div>
<div id="mcqs-advanced-linux" class="section level2" number="12.7">
<h2><span class="header-section-number">12.7</span> MCQs: Advanced linux</h2>
<p><img src="figures/question_bubble_red.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Please attempt to answer the below Multiple-Choice Questions to reinforce what you have learnt in this chapter.</p>
<ol style="list-style-type: decimal">
<li>Which symbol is used to pipe the output of one command to the input of another?
<div id="radio_AMTZLSELHU" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_AMTZLSELHU" value=""></input> <span><strong><code>&gt;</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_AMTZLSELHU" value="answer"></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_AMTZLSELHU" value=""></input> <span><strong><code>sed</code></strong></span></label>
</div></li>
<li>What command can be used to substitute text with other text?
<div id="radio_IVAQTYPKZL" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_IVAQTYPKZL" value=""></input> <span><strong><code>&gt;</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_IVAQTYPKZL" value=""></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_IVAQTYPKZL" value="answer"></input> <span><strong><code>sed</code></strong></span></label>
</div></li>
<li>Which symbol is used to redirect the output of a command to a file?
<div id="radio_QBJGYUOGBB" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_QBJGYUOGBB" value="answer"></input> <span><strong><code>&gt;</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_QBJGYUOGBB" value=""></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_QBJGYUOGBB" value=""></input> <span><strong><code>sed</code></strong></span></label>
</div></li>
<li>Which wildcard represents a single character?
<div id="radio_SVJRAKEFMZ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_SVJRAKEFMZ" value=""></input> <span><strong><code>*</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SVJRAKEFMZ" value=""></input> <span><strong><code>[]</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SVJRAKEFMZ" value="answer"></input> <span><strong><code>?</code></strong></span></label>
</div></li>
<li>Which wildcard represents zero or more characters?
<div id="radio_STEIJMITGP" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_STEIJMITGP" value="answer"></input> <span><strong><code>*</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_STEIJMITGP" value=""></input> <span><strong><code>[]</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_STEIJMITGP" value=""></input> <span><strong><code>?</code></strong></span></label>
</div></li>
<li>Which wildcard represents a range of characters?
<div id="radio_LWPOSYIIBP" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LWPOSYIIBP" value=""></input> <span><strong><code>*</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LWPOSYIIBP" value="answer"></input> <span><strong><code>[]</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_LWPOSYIIBP" value=""></input> <span><strong><code>?</code></strong></span></label>
</div></li>
<li>Which regular expression indicates that the preceding character matches 1 or 0 times?
<div id="radio_ZSCJFILCJC" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ZSCJFILCJC" value=""></input> <span><strong><code>.</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZSCJFILCJC" value="answer"></input> <span><strong><code>?</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZSCJFILCJC" value=""></input> <span><strong><code>*</code></strong></span></label>
</div></li>
<li>Which regular expression represents a single character?
<div id="radio_WRGQBVHFFC" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_WRGQBVHFFC" value="answer"></input> <span><strong><code>.</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_WRGQBVHFFC" value=""></input> <span><strong><code>?</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_WRGQBVHFFC" value=""></input> <span><strong><code>*</code></strong></span></label>
</div></li>
<li>Which regular expression indicates that the preceding character matches zero or more times?
<div id="radio_SPQULXKZGX" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_SPQULXKZGX" value=""></input> <span><strong><code>.</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SPQULXKZGX" value=""></input> <span><strong><code>?</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SPQULXKZGX" value="answer"></input> <span><strong><code>*</code></strong></span></label>
</div></li>
<li>Which regular expression matches the end of the line?
<div id="radio_SOHOALEBQX" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_SOHOALEBQX" value=""></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SOHOALEBQX" value=""></input> <span><strong><code>^</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_SOHOALEBQX" value="answer"></input> <span><strong><code>$</code></strong></span></label>
</div></li>
<li>Which regular expression is a logical OR operator?
<div id="radio_DSXKOWMAOR" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_DSXKOWMAOR" value="answer"></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_DSXKOWMAOR" value=""></input> <span><strong><code>^</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_DSXKOWMAOR" value=""></input> <span><strong><code>$</code></strong></span></label>
</div></li>
<li>Which regular expression matches the start of the line?
<div id="radio_DPLVRCDXGT" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_DPLVRCDXGT" value=""></input> <span><strong><code>|</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_DPLVRCDXGT" value="answer"></input> <span><strong><code>^</code></strong></span></label><label><input type="radio" autocomplete="off" name="radio_DPLVRCDXGT" value=""></input> <span><strong><code>$</code></strong></span></label>
</div></li>
</ol>

</div>
</div>
<div id="advance-practice-exercise" class="section level1" number="13">
<h1><span class="header-section-number">Chapter 13</span> Advance practice exercise</h1>
<p><img src="figures/adv_exercise.png" width="20%" style="display: block; margin: auto;" /></p>
<ol style="list-style-type: decimal">
<li>Copy the directory <strong>~/Linux/advanced_practice</strong> to <strong>~/Linux/advanced_practice_exercise</strong></li>
<li>Move into <strong>~/Linux/advanced_practice_exercise</strong></li>
<li>Make a directory called <strong>fastq</strong> and one called <strong>txt</strong></li>
<li>With one command move all the fastq files into the directory <strong>fastq</strong></li>
<li>With one command move all the txt files, excluding <strong>metadata.txt</strong> and <strong>samples.txt</strong>, into the directory <strong>txt</strong></li>
<li>Create a file in the <strong>fastq</strong> directory called <strong>patient_1_corrected.fastq</strong> and put all the corrected fastq data for patient_1 into the file. You can look at the metadat.txt file to see which samples belong to patient_1.</li>
<li>Append the metadata line for sample_1_AAAA to the bottom of the file <strong>sample_1_AAAA.txt</strong> in the <strong>txt</strong> directory.</li>
<li>For all the corrected fastq files find the sequences that start with a stop codon in the forward orientation (i.e. TAG, TAA or TGA). Print out to screen the sample name and sequence info separated by a “:” only (e.g. sample_10_AAGT:TAAGAGAACAATGAACAGATATTAATAATTTTGCCGCTTTTCTGCGGGAT)</li>
<li>Count the number of Gs and Cs within file <strong>sample_16_AACC.fastq</strong></li>
<li>Get the fastq headers of sequences with homopolymers made of As with a length of 5 or greater for the uncorrected fastq files for samples 3,4,5,13,14 and 15 with one command.</li>
</ol>
<p>You can check my solutions in the <a href="#exerciseadv_answers">Answers section</a>. These are not the definitive solution but only examples of solutions. If your method works and you understand why then you have done it correctly.</p>

</div>
<div id="bfxlanguages" class="section level1" number="14">
<h1><span class="header-section-number">Chapter 14</span> Bioinformatic Languages</h1>
<p><img src="figures/languages.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Unfortunately you cannot do everything you would want directly on the Linux command line. Even the tasks you can do are sometimes not very efficient or easy.
Fortunately there are many other programming languages you can use. There is a large amount of other programming languages and it can be hard to know which one to learn. Below is a list of commonly used bioinformatic program languages with a brief summary on their purpose and links to online resources to introduce you to the language.</p>
<div id="awk" class="section level2" number="14.1">
<h2><span class="header-section-number">14.1</span> awk</h2>
<p><code>awk</code> is typically used as a data extraction and reporting tool. It is very good due to its power and versatility.</p>
<p>Tutorial:</p>
<p><a href="http://www.grymoire.com/Unix/Awk.html" class="uri">http://www.grymoire.com/Unix/Awk.html</a></p>
<p>Manual:</p>
<p><a href="https://www.gnu.org/software/gawk/manual/gawk.html" class="uri">https://www.gnu.org/software/gawk/manual/gawk.html</a></p>
</div>
<div id="python" class="section level2" number="14.2">
<h2><span class="header-section-number">14.2</span> Python</h2>
<p><img src="figures/python.png" width="20%" style="display: block; margin: auto;" />
Python is used for software development and other applications It is favoured in Bioinformatics due to its relative ease to learn and it is able to handle strings well (i.e. genetic code). There are also many packages for python that help the analysis of biological data and other scientific data.</p>
<p>Python website:</p>
<p><a href="https://www.python.org/" class="uri">https://www.python.org/</a></p>
<p>BioPython:</p>
<p><a href="https://biopython.org/" class="uri">https://biopython.org/</a></p>
</div>
<div id="perl" class="section level2" number="14.3">
<h2><span class="header-section-number">14.3</span> Perl</h2>
<p><img src="figures/perl.png" width="20%" style="display: block; margin: auto;" />
Perl is a similar language to python with similar uses. The main difference is how they look.</p>
<p>Tutorial:</p>
<p><a href="https://www.tutorialspoint.com/perl/perl_introduction.htm" class="uri">https://www.tutorialspoint.com/perl/perl_introduction.htm</a></p>
<p>Perl website:</p>
<p><a href="https://www.perl.org/" class="uri">https://www.perl.org/</a></p>
<p>BioPerl:</p>
<p><a href="https://bioperl.org/" class="uri">https://bioperl.org/</a></p>
</div>
<div id="python-or-perl" class="section level2" number="14.4">
<h2><span class="header-section-number">14.4</span> Python or Perl?</h2>
<p>Generally speaking people normally choose to either learn python or perl. However there are many bioinformatic tools written in perl and many written in python. Therefore it is good to know a bit in both so you can debug scripts and specialise in one so you can write your own scripts. Our suggestion would be to specialise in Python due to its ease to learn and its growing popularity in the Bioinformatics field. However there are also other upcoming programming languages that are becoming more popular in bioinformatics, so you may like to give them a look.</p>
</div>
<div id="ruby" class="section level2" number="14.5">
<h2><span class="header-section-number">14.5</span> Ruby</h2>
<p><img src="figures/ruby.png" width="20%" style="display: block; margin: auto;" />
A programming language that is becoming more popular in bioinformatics due to its beginner friendliness.</p>
<p>Tutorial:</p>
<p><a href="http://rubylearning.com/" class="uri">http://rubylearning.com/</a></p>
<p>Ruby website:</p>
<p><a href="https://www.ruby-lang.org/en/" class="uri">https://www.ruby-lang.org/en/</a></p>
<p>BioRuby:</p>
<p><a href="http://bioruby.org/" class="uri">http://bioruby.org/</a></p>
<p>Why learn Ruby?</p>
<p><a href="http://www.bestprogramminglanguagefor.me/why-learn-ruby" class="uri">http://www.bestprogramminglanguagefor.me/why-learn-ruby</a></p>
</div>
<div id="golang-aka-go" class="section level2" number="14.6">
<h2><span class="header-section-number">14.6</span> Golang (AKA Go)</h2>
<p><img src="figures/golang.png" width="20%" style="display: block; margin: auto;" />
Another easy programming language with a library specifically made for bioinformatics.</p>
<p>Tutorial:</p>
<p><a href="https://tour.golang.org/welcome/1" class="uri">https://tour.golang.org/welcome/1</a></p>
<p>Go website:</p>
<p><a href="https://golang.org/" class="uri">https://golang.org/</a></p>
<p>biogo:</p>
<p><a href="https://github.com/biogo/biogo" class="uri">https://github.com/biogo/biogo</a></p>
</div>
<div id="r" class="section level2" number="14.7">
<h2><span class="header-section-number">14.7</span> R</h2>
<p><img src="figures/R.png" width="20%" style="display: block; margin: auto;" />
R is a very powerful programming language for statistical analysis and visualisation. It unfortunately has a large barrier to entry and normally quite unclear documentation. However it will unlikely be surpassed by another language any time soon due to its widespread use and large amount of very useful and powerful packages for various uses in the public domain. We would recommend using the IDE Rstudio when using R.</p>
<p>Tutorial:</p>
<p><a href="https://swirlstats.com/" class="uri">https://swirlstats.com/</a></p>
<p>R Website:</p>
<p><a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a></p>
<p>Cran Website:</p>
<p><a href="https://cran.r-project.org/" class="uri">https://cran.r-project.org/</a></p>
<p>Rstudio Website:</p>
<p><a href="https://www.rstudio.com/" class="uri">https://www.rstudio.com/</a></p>

</div>
</div>



<div id="exercise1_answers" class="section level1 unnumbered">
<h1>Exercise 1</h1>
<p><img src="figures/exercise.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Ensure you are in the correct directory before carrying out the below commands</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb104-1"><a href="#cb104-1" tabindex="-1"></a><span class="bu">cd</span> ~/Linux/</span></code></pre></div>
<ol style="list-style-type: decimal">
<li>Change the name of the subdirectory 4_exercises_aaaa within your Linux to 4_exercises</li>
</ol>
<div class="sourceCode" id="cb105"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb105-1"><a href="#cb105-1" tabindex="-1"></a><span class="fu">mv</span> 4_exercises_aaaa 4_exercises</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Make a backup of the 4_exercises directory</li>
</ol>
<div class="sourceCode" id="cb106"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb106-1"><a href="#cb106-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> 4_exercises 4_exercises_backup </span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>List the contents of the 4_exercises directory</li>
</ol>
<div class="sourceCode" id="cb107"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb107-1"><a href="#cb107-1" tabindex="-1"></a><span class="fu">ls</span> 4_exercises</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Within the directory 4_exercises</li>
</ol>
<div class="sourceCode" id="cb108"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb108-1"><a href="#cb108-1" tabindex="-1"></a><span class="bu">cd</span> 4_exercises</span></code></pre></div>
<p>Print the working directory</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb109-1"><a href="#cb109-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre></div>
<ol start="2" style="list-style-type: lower-alpha">
<li>Print out to screen the phrase ‘the echo command allows me to print phrases to screen’</li>
</ol>
<div class="sourceCode" id="cb110"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb110-1"><a href="#cb110-1" tabindex="-1"></a><span class="bu">echo</span> “the echo command allows me to print phrases to screen”</span></code></pre></div>
<ol start="3" style="list-style-type: lower-alpha">
<li>Copy the file copy_this_file.txt to the directory to_me</li>
</ol>
<div class="sourceCode" id="cb111"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb111-1"><a href="#cb111-1" tabindex="-1"></a><span class="fu">cp</span> copy_this_file.txt to_me</span></code></pre></div>
<ol start="4" style="list-style-type: lower-alpha">
<li>Rename the directory to_me to you</li>
</ol>
<div class="sourceCode" id="cb112"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb112-1"><a href="#cb112-1" tabindex="-1"></a><span class="fu">mv</span> to_me you</span></code></pre></div>
<ol start="6" style="list-style-type: lower-alpha">
<li>Delete the initial copy_this_file.txt file</li>
</ol>
<div class="sourceCode" id="cb113"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb113-1"><a href="#cb113-1" tabindex="-1"></a><span class="fu">rm</span> copy_this_file.txt</span></code></pre></div>
</div>
<div id="exercise2_answers" class="section level1 unnumbered">
<h1>Exercise 2</h1>
<p><img src="figures/exercise_2.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Ensure you are in the correct directory before carrying out the below commands</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb114-1"><a href="#cb114-1" tabindex="-1"></a><span class="bu">cd</span> ~/Linux/6_final_exercise</span></code></pre></div>
<ol style="list-style-type: decimal">
<li>See what files are in the directory</li>
</ol>
<div class="sourceCode" id="cb115"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb115-1"><a href="#cb115-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Rename the file 3-P£_CACTTCGA_L001_R1_001.fastq to 3-P3_CACTTCGA_L001_R1_001.fastq</li>
</ol>
<div class="sourceCode" id="cb116"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb116-1"><a href="#cb116-1" tabindex="-1"></a><span class="fu">mv</span> 3-P£_CACTTCGA_L001_R1_001.fastq <span class="dt">\</span></span>
<span id="cb116-2"><a href="#cb116-2" tabindex="-1"></a>3-P3_CACTTCGA_L001_R1_001.fastq</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Make a backup of the files in a directory called backup</li>
</ol>
<div class="sourceCode" id="cb117"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb117-1"><a href="#cb117-1" tabindex="-1"></a><span class="fu">mkdir</span> backup</span>
<span id="cb117-2"><a href="#cb117-2" tabindex="-1"></a><span class="fu">cp</span> 1-P1_ATGCCTGG_L001_R1_001.fastq backup/</span>
<span id="cb117-3"><a href="#cb117-3" tabindex="-1"></a><span class="fu">cp</span> 1-P1_ATGCCTGG_L001_R2_001.fastq backup/</span>
<span id="cb117-4"><a href="#cb117-4" tabindex="-1"></a><span class="fu">cp</span> 2-P2_AAGGACAC_L001_R1_001.fastq backup/</span>
<span id="cb117-5"><a href="#cb117-5" tabindex="-1"></a><span class="fu">cp</span> 2-P2_AAGGACAC_L001_R2_001.fastq backup/</span>
<span id="cb117-6"><a href="#cb117-6" tabindex="-1"></a><span class="fu">cp</span> 3-P3_CACTTCGA_L001_R1_001.fastq backup/</span>
<span id="cb117-7"><a href="#cb117-7" tabindex="-1"></a><span class="fu">cp</span> 3-P3_CACTTCGA_L001_R2_001.fastq backup/</span>
<span id="cb117-8"><a href="#cb117-8" tabindex="-1"></a><span class="fu">cp</span> 4-E1_ATTGGCTC_L001_R1_001.fastq backup/</span>
<span id="cb117-9"><a href="#cb117-9" tabindex="-1"></a><span class="fu">cp</span> 4-E1_ATTGGCTC_L001_R2_001.fastq backup/</span>
<span id="cb117-10"><a href="#cb117-10" tabindex="-1"></a><span class="fu">cp</span> metadata.txt backup/</span></code></pre></div>
<p>This can be done a lot quicker with the use of wildcard characters (Covered in Advanced Linux section)</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb118-1"><a href="#cb118-1" tabindex="-1"></a><span class="fu">mkdir</span> backup</span>
<span id="cb118-2"><a href="#cb118-2" tabindex="-1"></a><span class="fu">cp</span> <span class="pp">*</span>fastq backup</span>
<span id="cb118-3"><a href="#cb118-3" tabindex="-1"></a><span class="fu">cp</span> <span class="pp">*</span>txt backup</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>How many reads are in the samples?
The below command will give the number of lines in the files, this number can then be divided by 4 (mentally or using a calculator). These values will be the same for the R2 files as they are for the matching R1 file.</li>
</ol>
<div class="sourceCode" id="cb119"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb119-1"><a href="#cb119-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> 1-P1_ATGCCTGG_L001_R1_001.fastq <span class="dt">\</span></span>
<span id="cb119-2"><a href="#cb119-2" tabindex="-1"></a>2-P2_AAGGACAC_L001_R1_001.fastq <span class="dt">\</span></span>
<span id="cb119-3"><a href="#cb119-3" tabindex="-1"></a>3-P3_CACTTCGA_L001_R1_001.fastq</span></code></pre></div>
<p>An advanced method using regular expressions, wildcard characters and grep</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb120-1"><a href="#cb120-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">&quot;^@[0-9]*_&quot;</span> <span class="pp">*</span>R1<span class="pp">*</span>.fastq</span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>Remove the fastq files with no data
Check which files have no data</li>
</ol>
<div class="sourceCode" id="cb121"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb121-1"><a href="#cb121-1" tabindex="-1"></a><span class="fu">wc</span> <span class="dt">\</span></span>
<span id="cb121-2"><a href="#cb121-2" tabindex="-1"></a>1-P1_ATGCCTGG_L001_R1_001.fastq 1-P1_ATGCCTGG_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb121-3"><a href="#cb121-3" tabindex="-1"></a>2-P2_AAGGACAC_L001_R1_001.fastq 2-P2_AAGGACAC_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb121-4"><a href="#cb121-4" tabindex="-1"></a>3-P3_CACTTCGA_L001_R1_001.fastq 3-P3_CACTTCGA_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb121-5"><a href="#cb121-5" tabindex="-1"></a>4-E1_ATTGGCTC_L001_R1_001.fastq 4-E1_ATTGGCTC_L001_R2_001.fastq </span></code></pre></div>
<p>Remove empty files</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb122-1"><a href="#cb122-1" tabindex="-1"></a><span class="fu">rm</span> <span class="dt">\</span></span>
<span id="cb122-2"><a href="#cb122-2" tabindex="-1"></a>4-E1_ATTGGCTC_L001_R1_001.fastq 4-E1_ATTGGCTC_L001_R2_001.fastq</span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>Update the backup files with the previous change</li>
</ol>
<div class="sourceCode" id="cb123"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb123-1"><a href="#cb123-1" tabindex="-1"></a><span class="fu">rm</span> backup/4-E1_ATTGGCTC_L001_R1_001.fastq <span class="dt">\</span></span>
<span id="cb123-2"><a href="#cb123-2" tabindex="-1"></a>backup/4-E1_ATTGGCTC_L001_R2_001.fastq </span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>Check if the 1st read names match in the paired files</li>
</ol>
<div class="sourceCode" id="cb124"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb124-1"><a href="#cb124-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 1 <span class="dt">\</span></span>
<span id="cb124-2"><a href="#cb124-2" tabindex="-1"></a>1-P1_ATGCCTGG_L001_R1_001.fastq 1-P1_ATGCCTGG_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb124-3"><a href="#cb124-3" tabindex="-1"></a>2-P2_AAGGACAC_L001_R1_001.fastq 2-P2_AAGGACAC_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb124-4"><a href="#cb124-4" tabindex="-1"></a>3-P3_CACTTCGA_L001_R1_001.fastq 3-P3_CACTTCGA_L001_R2_001.fastq </span></code></pre></div>
<ol start="8" style="list-style-type: decimal">
<li>Check if the last read names match in the paired files</li>
</ol>
<div class="sourceCode" id="cb125"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb125-1"><a href="#cb125-1" tabindex="-1"></a><span class="fu">tail</span> <span class="at">-n</span> 4 <span class="dt">\</span></span>
<span id="cb125-2"><a href="#cb125-2" tabindex="-1"></a>1-P1_ATGCCTGG_L001_R1_001.fastq 1-P1_ATGCCTGG_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb125-3"><a href="#cb125-3" tabindex="-1"></a>2-P2_AAGGACAC_L001_R1_001.fastq 2-P2_AAGGACAC_L001_R2_001.fastq <span class="dt">\</span></span>
<span id="cb125-4"><a href="#cb125-4" tabindex="-1"></a>3-P3_CACTTCGA_L001_R1_001.fastq 3-P3_CACTTCGA_L001_R2_001.fastq </span></code></pre></div>
<ol start="9" style="list-style-type: decimal">
<li>In file 1-P1_ATGCCTGG_L001_R1_001.fastq look for sequence headers with the term ‘psychrobacter’</li>
</ol>
<div class="sourceCode" id="cb126"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb126-1"><a href="#cb126-1" tabindex="-1"></a><span class="fu">grep</span> “psychrobacter” 1-P1_ATGCCTGG_L001_R1_001.fastq</span></code></pre></div>
<ol start="10" style="list-style-type: decimal">
<li>In the sample 1-P1 remove any fastq entries where the term ‘psychrobacter’ appears in the fastq header. Do this for the R1 and R2 file.
<ul>
<li>Using <strong>nano</strong> use “Ctrl+W” to search for psychrobacter. Then use “Ctrl+K” to cut the lines followed by “Ctrl+W” and “Ctrl+X” to save and exit.</li>
<li>Using <strong>vim</strong> with “/” to search for psychrobacter, “dd” to delete lines and “:wq” to save the file and exit it.</li>
</ul></li>
<li>Print to screen the fastq header, sequence and quality data for the 25th sequence in sample 2-P2 for both the R1 and R2 file. Do this with one command.
@24_ecoli is grepped as the first sequence is @0_ecoli</li>
</ol>
<div class="sourceCode" id="cb127"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb127-1"><a href="#cb127-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-A</span> 3 <span class="st">&quot;@24_ecoli&quot;</span> 2-P2_AAGGACAC_L001_R1_001.fastq</span>
<span id="cb127-2"><a href="#cb127-2" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-A</span> 3 <span class="st">&quot;@24_ecoli&quot;</span> 2-P2_AAGGACAC_L001_R2_001.fastq</span></code></pre></div>
</div>
<div id="exerciseadv_answers" class="section level1 unnumbered">
<h1>Advanced exercise</h1>
<p><img src="figures/adv_exercise.png" width="20%" style="display: block; margin: auto;" />
1. Copy the directory ~/Linux/advanced_practice to ~/Linux/advanced_practice_exercise</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb128-1"><a href="#cb128-1" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ~/Linux/advanced_practice ~/Linux/advanced_practice_exercise</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Move into ~/Linux/advanced_practice_exercise</li>
</ol>
<div class="sourceCode" id="cb129"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb129-1"><a href="#cb129-1" tabindex="-1"></a><span class="bu">cd</span> ~/Linux/advanced_practice_exercise</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Make a directory called fastq and one called txt</li>
</ol>
<div class="sourceCode" id="cb130"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb130-1"><a href="#cb130-1" tabindex="-1"></a><span class="fu">mkdir</span> fastq txt</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>With one command move all the fastq files into the directory fastq</li>
</ol>
<div class="sourceCode" id="cb131"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb131-1"><a href="#cb131-1" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.fastq fastq/</span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>With one command move all the txt files, excluding metadata.txt and samples.txt, into the directory txt.</li>
</ol>
<div class="sourceCode" id="cb132"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb132-1"><a href="#cb132-1" tabindex="-1"></a><span class="fu">mv</span> sample_<span class="pp">*</span>txt txt/</span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>Create a file in the fastq directory called patient_1_corrected.fastq and put all the corrected fastq data for patient_1 into the file.</li>
</ol>
<div class="sourceCode" id="cb133"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb133-1"><a href="#cb133-1" tabindex="-1"></a><span class="fu">cat</span> fastq/sample_<span class="pp">[</span><span class="ss">1</span><span class="pp">-</span><span class="ss">2</span><span class="pp">]</span>_<span class="pp">*</span>corrected.fastq <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb133-2"><a href="#cb133-2" tabindex="-1"></a>fastq/patient_1_corrected.fastq</span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>Append the metadata line for sample_1_AAAA from metadata.txt to the bottom of the file sample_1_AAA.txt in the txt directory.</li>
</ol>
<div class="sourceCode" id="cb134"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb134-1"><a href="#cb134-1" tabindex="-1"></a><span class="fu">cat</span> metadata.txt <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;sample_1_AAAA&quot;</span> <span class="op">&gt;&gt;</span> txt/sample_1_AAAA.txt</span></code></pre></div>
<ol start="8" style="list-style-type: decimal">
<li>For all the corrected fastq files find the sequences that start with a stop codon in the forward orientation (i.e. TAG, TAA or TGA). Print out to screen the sample name and sequence info separated by a “:” only (i.e. sample_10_AAGT:TAAGAGAACAATGAACAGATATTAATAATTTTGCCGCTTTTCTGCGGGAT)</li>
</ol>
<div class="sourceCode" id="cb135"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb135-1"><a href="#cb135-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">&quot;^TA[AG]\|^TGA&quot;</span> fastq/<span class="pp">*</span>corrected.fastq <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb135-2"><a href="#cb135-2" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&quot;s/.*sample/sample/&quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/_corrected.fastq//&quot;</span></span></code></pre></div>
<ol start="9" style="list-style-type: decimal">
<li>Count the number of Gs and Cs within the sequences of file sample_16_AACC.fastq</li>
</ol>
<div class="sourceCode" id="cb136"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb136-1"><a href="#cb136-1" tabindex="-1"></a><span class="fu">cat</span> fastq/sample_16_AACC.fastq <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-B</span> 1 <span class="st">&quot;^+$&quot;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb136-2"><a href="#cb136-2" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">&quot;+\|--&quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/A\|T//g&quot;</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-c</span></span></code></pre></div>
<ol start="10" style="list-style-type: decimal">
<li>Get the fastq headers of sequences with homopolymers made of As with a length of 5 or greater for the uncorrected fastq files for samples 8-13 with one command. Then in the same command make the final output of each line in the format of “Sample_13: Sequence 12”</li>
</ol>
<div class="sourceCode" id="cb137"><pre class="sourceCode bash Rchunk"><code class="sourceCode bash"><span id="cb137-1"><a href="#cb137-1" tabindex="-1"></a><span class="fu">cat</span> fastq/<span class="pp">*[</span><span class="ss">3</span><span class="pp">-</span><span class="ss">5</span><span class="pp">]*[</span><span class="ss">AGCT</span><span class="pp">]</span>.fastq <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-B</span> 2 <span class="st">&quot;^+$&quot;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb137-2"><a href="#cb137-2" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B</span> 1 <span class="st">&quot;AAAAA&quot;</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;^@&quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/^@s/S/&quot;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb137-3"><a href="#cb137-3" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&quot;s/_[AGCT]*_/: Sequence /&quot;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&quot;s/ 1:$//&quot;</span></span></code></pre></div>

</div>



<div id="cheatsheet" class="section level1" number="15">
<h1><span class="header-section-number">A</span> Practical cheat sheet</h1>
<p><img src="figures/cheatsheet.png" width="20%" style="display: block; margin: auto;" /></p>
<table style="width:99%;">
<colgroup>
<col width="11%" />
<col width="40%" />
<col width="48%" />
</colgroup>
<thead>
<tr class="header">
<th><strong>Command</strong></th>
<th><strong>Description</strong></th>
<th><strong>Usage example</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>echo</td>
<td>Print out a string to screen</td>
<td>Echo ‘Hello World’</td>
</tr>
<tr class="even">
<td>man</td>
<td>Look at the manual page of a command</td>
<td>man man</td>
</tr>
<tr class="odd">
<td>clear</td>
<td>Clear the screen</td>
<td>clear</td>
</tr>
<tr class="even">
<td>&lt;tab&gt;</td>
<td>Complete a path or command</td>
<td>&lt;tab&gt;</td>
</tr>
<tr class="odd">
<td>cd</td>
<td>Change directory</td>
<td>cd /directory/path/</td>
</tr>
<tr class="even">
<td>pwd</td>
<td>Print Working Directory</td>
<td>pwd</td>
</tr>
<tr class="odd">
<td>ls</td>
<td>List the contents of a directory</td>
<td>ls /directory/path/</td>
</tr>
<tr class="even">
<td>mkdir</td>
<td>Make a directory</td>
<td>mkdir /directory/path/new_directory/</td>
</tr>
<tr class="odd">
<td>cp</td>
<td>Copy content to another path</td>
<td>cp /directory/path/file.txt /directory/path/new_directory/</td>
</tr>
<tr class="even">
<td>mv</td>
<td>Move content to a new path</td>
<td>mv /directory/path/file.txt /directory/path/new_directory/</td>
</tr>
<tr class="odd">
<td>rm</td>
<td>Delete content</td>
<td>rm /directory/path/file.txt</td>
</tr>
<tr class="even">
<td>cat</td>
<td>Print contents of a file to screen</td>
<td>cat /directory/path/file.txt</td>
</tr>
<tr class="odd">
<td>head</td>
<td>Print out the first n lines of a file to screen</td>
<td>head -n 10 /directory/path/file.txt</td>
</tr>
<tr class="even">
<td>tail</td>
<td>Print the last n lines of a file to screen</td>
<td>tail -n 12 /directory/path/file.txt</td>
</tr>
<tr class="odd">
<td>less</td>
<td>Read a file one page at a time</td>
<td>less /directory/path/file.txt</td>
</tr>
<tr class="even">
<td>wc</td>
<td>Print out line, word and byte count</td>
<td>wc /directory/path/file.txt</td>
</tr>
<tr class="odd">
<td>grep</td>
<td>Search for lines that contain a specific pattern</td>
<td>grep “pattern” /directory/path/file.txt</td>
</tr>
<tr class="even">
<td>vim</td>
<td>Text editor</td>
<td>vim /directory/path/file.txt</td>
</tr>
</tbody>
</table>
</div>
<div id="windows-terminals" class="section level1" number="16">
<h1><span class="header-section-number">B</span> Windows terminals</h1>
<p><img src="figures/windows.png" width="15%" style="display: block; margin: auto;" /></p>
<p>If you are on a windows you may need to download a terminal program to <code>ssh</code> to a linux cluster. Thankfully Macs come with an in built terminal. Below are a few suggestions of windows terminals:</p>
<ul>
<li><a href="https://mobaxterm.mobatek.net/">MobaXterm</a></li>
<li><a href="https://www.putty.org/">Putty</a></li>
<li><a href="https://www.microsoft.com/en-us/p/ubuntu/9nblggh4msv6?activetab=pivot:overviewtab">Ubuntu on Windows</a></li>
</ul>
</div>
<div id="ssh" class="section level1" number="17">
<h1><span class="header-section-number">C</span> <code>ssh</code></h1>
<p><img src="figures/turtle_shell.png" width="15%" style="display: block; margin: auto;" /></p>
<p>The <code>ssh</code> (Secure Shell Protocol) command is used to login into cluster and other machines. For more information on this please see the following online tutorial: <a href="https://opensource.com/article/20/9/ssh" class="uri">https://opensource.com/article/20/9/ssh</a></p>
</div>
<div id="file-transferring" class="section level1" number="18">
<h1><span class="header-section-number">D</span> File transferring</h1>
<p><img src="figures/file_transfer.png" width="15%" style="display: block; margin: auto;" /></p>
<p>When working with a remote cluster you will most likely want to transfer files from your computer to the cluster and vice versa.</p>
<p>For transferring on windows machine I would suggest <a href="https://winscp.net/eng/index.php">WinSCP</a>.</p>
<p>For transferring on Macs I would suggest <a href="https://filezilla-project.org/">FileZIlla</a></p>
</div>
<div id="mamba" class="section level1" number="19">
<h1><span class="header-section-number">E</span> Mamba</h1>
<p><img src="figures/mamba_logo.png" width="25%" style="display: block; margin: auto;" /></p>
<div id="mamba_install" class="section level2" number="19.1">
<h2><span class="header-section-number">E.1</span> Mamba installation</h2>
<p>Once you start using bioinformatic tools you will notice that a lot of installing is required. I would highly suggest using Mamba for this purpose.</p>
<p>Mamba is a reimplementation of conda. It is a great tool for installing bioinformatic packages including R packages.</p>
<p>Mamba github: <a href="https://github.com/mamba-org/mamba" class="uri">https://github.com/mamba-org/mamba</a></p>
<p>The best way to use Mamba is to install Miniforge. It has both Conda and Mamba commands.</p>
<p>Miniforge installation: <a href="https://github.com/conda-forge/miniforge" class="uri">https://github.com/conda-forge/miniforge</a></p>
<p>Mamba guide: <a href="https://mamba.readthedocs.io/en/latest/user_guide/mamba.html" class="uri">https://mamba.readthedocs.io/en/latest/user_guide/mamba.html</a></p>
</div>
</div>
<div id="vim" class="section level1" number="20">
<h1><span class="header-section-number">F</span> vim</h1>
<p><img src="figures/Vimlogo.png" width="20%" style="display: block; margin: auto;" />
To enter the <code>vim</code> text editor you can use the command <code>vim</code>. The command is: <code>vim file.txt</code>.</p>
<p><code>vim</code> can be run with a previous file name which you can then edit or a new file name in which case you will create a new file.</p>
<p>Once you are in the <code>vim</code> editor there are two main modes:</p>
<ol style="list-style-type: decimal">
<li><strong>Command mode</strong>: This is the starting mode of <code>vim</code>. It can be used to enter commands but not for typing.</li>
<li><strong>Insert mode</strong>: This mode can be used to insert characters into the text. Pasting only works properly in insert mode. You will know you are in insert mode as at the bottom of the screen will be <strong>“– INSERT –”</strong>.
<ol style="list-style-type: decimal">
<li>To <strong>enter</strong> insert mode press <strong>i</strong></li>
<li>To <strong>escape</strong> Insert mode press <strong>esc</strong></li>
</ol></li>
</ol>
<p>Below are a subset of commands you can use in the command mode of vim.</p>
<ul>
<li><strong>:q</strong> - Quit vim. Will fail if there are any unsaved changes</li>
<li><strong>:w</strong> - Save the file</li>
<li><strong>:wq</strong> - Save and quit</li>
<li><strong>:x</strong> - Save and quit</li>
<li><strong>:q!</strong> - Quit and throw away unsaved changes</li>
<li><strong>:saveas file</strong> - Save file as. File name can be changed</li>
<li><strong>:w file</strong> - Save file as. File name can be changed</li>
<li><strong>arrow keys</strong> - Navigate the text. (works in insert mode too)</li>
<li><strong>^</strong> - Jump to the start of a line</li>
<li><strong>$</strong> - Jump to the end of a line</li>
<li><strong>/pattern</strong> - Search for pattern</li>
<li><strong>n</strong> - Repeat search in same direction</li>
<li><strong>N</strong> - Repeat search in reverse direction</li>
<li><strong>dd</strong> - Delete the current line</li>
<li><strong>i</strong> - Enter insert mode</li>
<li><strong>1G</strong> - Go to the first line in the file. Number can be changed</li>
<li><strong>G</strong> - Go to the last line of the file</li>
</ul>
<p><a href="https://vim.rtorr.com/"><strong>Vim cheatsheet</strong></a></p>

</div>
<p style="text-align: center;">
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
